---
title: "Using NicheNet with OmnipathR"
author:
  - Denes Turei:
      affiliation: unihd
      email: turei.denes@gmail.com
      correspondance: true
  - name: Alberto Valdeolivas
    affiliation: unihd
  - name: Julio Saez-Rodriguez
    affiliation: unihd
institute:
  - unihd: Institute for Computational Biomedicine, Heidelberg University
package: OmnipathR
output:
  BiocStyle::html_document:
  pdf_document:
    number_sections: yes
    toc: yes
    toc_depth: 4
    pandoc_args:
      - '--lua-filter=scholarly-metadata.lua'
      - '--lua-filter=author-info-blocks.lua'
abstract: |
  NicheNet is a method to infer ligand activities from transcriptomics data.
  It relies on prior knowledge which includes signaling networks and
  transcriptomics data from ligand perturbation experiments. OmnipathR
  provides a close integration with NicheNet: it provides methods to build
  the networks directly from the original sources in a format suitable for
  NicheNet in a highly customizable way. OmnipathR also provides a workflow
  template which connects OmnipathR and NicheNet methods from building the
  prior knowledge up until the inference of ligand activities. With this
  pipeline the only thing users need to provide is processed transcriptomics
  data.
vignette: |
  %\VignetteIndexEntry{Using NicheNet with OmnipathR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
fig_width: 9
fig_height: 7
---

# Introduction

The OmnipathR package contains methods to build NichNet[1] format networks
from OmniPath[2] data and from the resources used in the original NicheNet
study. With a few exceptions, almost all resources used originally by
NicheNet are available by OmnipathR both in their raw format or after minimal
preprocessing, and in the format suitable for NicheNet. However, the networks
built by OmnipathR are not completely identical with the ones used by
NicheNet due to the complexity of processing such a large amount of data.
The advantage of building these networks with OmnipathR is the transparent
and reproducible process, as the origin of any data record can be easily
traced back to its original source. The data used in the NicheNet publication
is deposited at [Zenodo](https://zenodo.org/record/3260758). The other part
of the prior knowledge used by NicheNet is a collection of expression data
from ligand perturbation experiments. In OmnipathR This is taken directly
from the Zenodo repository.

Apart from building the prior knowledge, OmnipathR provides a number of glue
methods which connect elements of the NicheNet workflow, ultimately making it
possible to run everything with a single function call, starting from the
processed transcriptomics data and going until the predicted ligand
activities. To implement such a workflow we followed the [analysis from the
case study](https://workflows.omnipathdb.org/nichenet1.html) in the [OmniPath
paper](https://www.biorxiv.org/content/10.1101/2020.08.03.221242v2) and the
[vignettes](https://github.com/saeyslab/nichenetr/tree/master/vignettes) from
NicheNet. The workflow consists of thin wrappers around some of the NicheNet
methods. Parameters for these methods can be provided from the top level call.

The transcriptomics data is the main input for the pipeline which is specific
for the study. The transcriptomics data can be processed by standard tools,
such as
[DESeq](https://bioconductor.org/packages/release/bioc/html/DESeq.html)[3]
for bulk and [Seurat](https://satijalab.org/seurat/)[4] for single-cell
transcriptomics. For the latter more guidance is available in the [NicheNet
vignettes](https://github.com/saeyslab/nichenetr/blob/master/vignettes/
seurat_wrapper.md). The pipeline presented here only requires the list of
genes expressed in the transmitter and receiver cells (in case of autocrine
signaling these are the same); and a list with genes of interest. Genes of
interest might come from the investigation subject or gene set enrichment
analysis (see an example
[here](https://workflows.omnipathdb.org/nichenet1.html)).

# Run the workflow by a single call

The `nichenet_main` function executes all the workflow elements, hence it can
be used to run the complete workflow. Here we show it as an example, and in
the next sections we take a closer look at the individual steps. The first
four arguments comes from the study data and objective, as discussed above.
All parameters of the workflow can be overridden here, for example, for the
`mlrbo_optimization` we override the `ncores` argument, and exclude CPDB
from the signaling network and set the confidence threshold for EVEX.

```{r eval=FALSE}
library(OmnipathR)
library(nichenetr)

nichenet_results <- nichenet_main(
    expressed_genes_transmitter = expressed_genes_transmitter,
    expressed_genes_receiver = expressed_genes_receiver,
    genes_of_interest = genes_of_interest,
    background_genes = background_genes,
    signaling_network = list(
        cpdb = NULL,
        evex = list(min_confidence = 1.0)
    ),
    gr_network = list(only_omnipath = TRUE),
    n_top_ligands = 20,
    mlrmbo_optimization_param = list(ncores = 4)
)
```

# Only model building

Without transcriptomics data, it is possible to build a NicheNet model which
later can be used for the ligand activity prediction. Hence if the first four
arguments above are `NULL` the pipeline will run only the model building.
Still, in this tutorial we wrap all blocks into `eval = FALSE` because the
model optimization is computationally intensive, it might take hours to run.
To create the NicheNet model it's enough to call `nichenet_main` without
arguments, all further arguments are optional override of package defaults.

```{r eval=FALSE}
nichenet_model <- nichenet_main()
```

# Workflow steps

## Networks

NicheNet requires three types of interactions: signaling, ligand-receptor (lr)
and gene regulatory (gr). All these are collected from various resources. In
OmnipathR a top level function, `nichenet_networks` manages the build of all
the three networks. It returns a list with three data frames:

```{r eval=FALSE}
networks <- nichenet_networks()
```

Its `only_omnipath` argument is a shortcut to build all networks by using
only OmniPath data:

```{r eval=FALSE}
networks <- nichenet_networks(only_omnipath = TRUE)
```

To the network building functions of OmniPath many further arguments can
be provided, at the end these are passed to
`import_post_translational_interactions`, `import_intercell_network` and
`import_transcriptional_interactions`, so we recommend to read the manual of
these functions. As an example, one can restrict the signaling network to
the resources SIGNOR and PhosphoSite, while for ligand-receptor use only
ligands and enzymes on the transmitter side and receptors and transporters
on the receiver side, while for gene regulatory network, use only the A
confidence level of DoRothEA:

```{r eval=FALSE}
networks <- nichenet_networks(
    only_omnipath = TRUE,
    signaling_network = list(
        omnipath = list(
            resources = c('SIGNOR', 'PhosphoSite')
        )
    ),
    lr_network = list(
        omnipath = list(
            transmitter_param = list(parent = c('ligand', 'secreted_enzyme')),
            receiver_param = list(parent = c('receptor', 'transporter'))
        )
    ),
    gr_network = list(
        omnipath = list(
            resources = 'DoRothEA',
            dorothea_levels = 'A'
        )
    )
)
```

### Signaling network

The function `nichenet_signaling_network` builds the signaling network from
all resources with the default settings. Each argument of this function is a
list of arguments for a single resource. If an argument is set to `NULL`, the
resource will be omitted. In the example below we set custom confidence score
thresholds for ConsensusPathDB and EVEX, while use no data from Pathway
Commons. In this case, the rest of the resources will be loaded with their
default parameters:

```{r eval=FALSE}
signaling_network <- nichenet_signaling_network(
    cpdb = list(
        complex_max_size = 1,
        min_score = .98
    ),
    evex = list(
        min_confidence = 2
    ),
    pathwaycommons = NULL
)
```

Currently the following signaling network resources are available by
OmnipathR: OmniPath, Pathway Commons, Harmonizome (PhosphoSite, KEA and
DEPOD), Vinayagam et al. 2011, ConsensusPathDB, EVEX and InWeb InBioMap.
Let's take a closer look on one of the resource specific methods. The name
of these functions all follow the same scheme, e.g. for EVEX it is:
`nichenet_signaling_network_evex`. From most of the resources we just load
and format the interactions, but a few of them accepts parameters, such as
confidence score thresholds or switches to include or exclude certain kind
of records:

```{r eval=FALSE}
evex_signaling <- nichenet_signaling_network_evex(top_confidence = .9)
```

The network format of NicheNet is very minimalistic, it contains only the
interacting partners and the resource. To investigate the data more in depth,
we recommend to look at the original data. The methods retrieving data from
resources all end by `_download` for foreign resources and start by `import_`
for OmniPath. For example, to access the EVEX data as it is provided by the
resource:

```{r eval=FALSE}
evex <- evex_download(remove_negatives = FALSE)
```

### Ligand-receptor interactions and gene regulation

These networks work exactly the same way as the signaling network: they are
built by the `nichenet_lr_network` and `nichenet_gr_network` functions.
Currently the following ligand-receptor network resources are available in
OmnipathR: OmniPath, Ramilowski et al. 2015, Guide to Pharmacology
(IUPHAR/BPS); and in the gene regulatory network: OmniPath, Harmonizome,
RegNetwork, HTRIdb, ReMap, EVEX, PathwayCommons and TRRUST.

## Ligand perturbation experiments

The expression data from the collection of over a hundred of ligand
perturbation experiments is available in the data deposited by NicheNet
authors in [Zenodo](https://zenodo.org/record/3260758), in the
`expression_settings.rds` file. The function below downloads and imports
this data:

```{r eval=FALSE}
expression <- nichenet_expression_data()
```

If there is any ligand which is missing from the ligand-receptor network,
probably we want to remove it:

```{r eval=FALSE}
lr_network <- nichenet_lr_network()
expression <- nichenet_remove_orphan_ligands(
    expression = expression,
    lr_network = lr_network
)
```

## Model optimization

In this step we optimize the parameters of the model and assign weights to
the network resources, especially based on the ligand perturbation data. The
`nichenet_optimization` function is a wrapper around
`nichenetr::mlrmbl_optimization`. The arguments for this function, and the
objective function can be overriden, for example it's a good idea to set
the cores according to our machine's CPU count:

```{r eval=FALSE}
optimization_results <- nichenet_optimization(
    networks = networks,
    expression = expression,
    mlrmbo_optimization_param = list(ncores = 4)
)
```

This function takes very long, even hours to run. Before it returns the
results, it saves them into an RDS file. This and the further RDS files are
saved into the `nichenet_results` directory by default, it can be changed
by the option `omnipath.nichenet_results_dir`.

```{r eval=FALSE}
options(omnipath.nichenet_results_dir = 'my/nichenet/dir')
nichenet_results_dir()
# [1] "my/nichenet/dir"
```

## Model build

The next major part is to build a model which connects ligands to targets,
i.e. which genes are affected in their expression by which of the ligands.
Again, we use a wrapper around NicheNet functions: the `nichenet_build_model`
calls `nichenetr::onstruct_weighted_networks` and
`nichenetr::apply_hub_corrections`. If the argument `weights` is `FALSE`,
the optimized resource weights are not used, all resources considered with
the same weight. The results are saved automatically into an RDS.

```{r eval=FALSE}
nichenet_model <- nichenet_build_model(
    optimization_results = optimization_results,
    networks = networks,
)
```

## Ligand-target matrix

Finally, we produce a ligand-target matrix, a sparse matrix with weighted
connections from ligands to targets. The function
`nichenet_ligand_target_matrix` is a wrapper around
`nichenetr::construct_ligand_target_matrix`. As usual the results are
exported to an RDS. This is the last step of the NicheNet model building
process.

```{r eval=FALSE}
lt_matrix <- nichenet_ligand_target_matrix(
    nichenet_model$weighted_networks,
    networks$lr_network,
    nichenet_model$optimized_parameters
)
```

## Ligand activities

In this step we ues the NicheNet model to infer ligand activities from our
transcriptomics data. The last three arguments are all character vectors
with gene symbols:

```{r eval=FALSE}
ligand_activities <- nichenet_ligand_activities(
    ligand_target_matrix = lt_matrix,
    lr_network = networks$lr_network,
    expressed_genes_transmitter = expressed_genes_transmitter,
    expressed_genes_receiver = expressed_genes_receiver,
    genes_of_interest = genes_of_interest
)
```

Once the ligand activities inferred, we can obtain a list of top ligand-target
links. It's up to us how many of the top ranking ligands, and for each of
these ligands, how many of their top targets we include in this table:

```{r eval=FALSE}
lt_links <- nichenet_ligand_target_links(
    ligand_activities = ligand_activities,
    ligand_target_matrix = lt_matrix,
    genes_of_interest = genes_of_interest,
    n_top_ligands = 20,
    n_top_targets = 100
)
```

# Further steps

The NicheNet workflow in OmnipathR is implemented until this point. The
results can be visualized by the methods included in `nichenetr`, can be
analysed for enrichment of functional properties, or the signaling network
connecting the ligands to targets can be reconstructed and analysed,
as it's shown in the OmniPath
[case study](https://workflows.omnipathdb.org/nichenet1.html).

# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References

[1] R Browaeys, W Saelens and Y Saeys (2020) NicheNet: modeling intercellular
communication by linking ligands to target genes. _Nat Methods_ 17, 159â€“162

[2] D Turei, A Valdeolivas, L Gul, N Palacio-Escat, O Ivanova, A Gabor,
D Modos, T Korcsmaros and J Saez-Rodriguez (2020) Integrated intra- and
intercellular signaling knowledge for multicellular omics analysis. _bioRxiv_
2020.08.03.221242

[3] S Anders, W Huber (2010) Differential expression analysis for sequence
count data. _Genome Biol_ 11, R106

[4] Y Hao, S Hao, E Andersen-Nissen, WM Mauck, S Zheng, A Butler, MJ Lee,
AJ Wilk, C Darby, M Zagar, P Hoffman, M Stoeckius, E Papalexi, EP Mimitou,
J Jain, A Srivastava, T Stuart, LB Fleming, B Yeung, AJ Rogers, JM McElrath,
CA Blish, R Gottardo, P Smibert and R Satija (2020) Integrated analysis of
multimodal single-cell data. _bioRxiv_ 2020.10.12.335331