<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>OmniPath web service client and more • OmnipathR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="OmniPath web service client and more">
<meta name="description" content="A client for the OmniPath web service (https://www.omnipathdb.org) and many other resources. It also includes functions to transform and pretty print some of the downloaded data, functions to access a number of other resources such as BioPlex, ConsensusPathDB, EVEX, Gene Ontology, Guide to Pharmacology (IUPHAR/BPS), Harmonizome, HTRIdb, Human Phenotype Ontology, InWeb InBioMap, KEGG Pathway, Pathway Commons, Ramilowski et al. 2015, RegNetwork, ReMap, TF census, TRRUST and Vinayagam et al. 2011. Furthermore, OmnipathR features a close integration with the NicheNet method for ligand activity prediction from transcriptomics data, and its R implementation `nichenetr` (available only on github).">
<meta property="og:description" content="A client for the OmniPath web service (https://www.omnipathdb.org) and many other resources. It also includes functions to transform and pretty print some of the downloaded data, functions to access a number of other resources such as BioPlex, ConsensusPathDB, EVEX, Gene Ontology, Guide to Pharmacology (IUPHAR/BPS), Harmonizome, HTRIdb, Human Phenotype Ontology, InWeb InBioMap, KEGG Pathway, Pathway Commons, Ramilowski et al. 2015, RegNetwork, ReMap, TF census, TRRUST and Vinayagam et al. 2011. Furthermore, OmnipathR features a close integration with the NicheNet method for ligand activity prediction from transcriptomics data, and its R implementation `nichenetr` (available only on github).">
<meta property="og:image" content="https://saezlab.github.io/OmnipathR/logo.svg">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-10"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-10');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">OmnipathR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.15.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/bioc_workshop.html">OmniPath Bioconductor workshop</a></li>
    <li><a class="dropdown-item" href="articles/cosmos.html">Building prior knowledge network (PKN) for COSMOS</a></li>
    <li><a class="dropdown-item" href="articles/db_manager.html">The database manager in OmnipathR</a></li>
    <li><a class="dropdown-item" href="articles/drug_targets.html">Building protein networks around drug-targets using OmnipathR</a></li>
    <li><a class="dropdown-item" href="articles/extra_attrs.html">Resource specific interaction attributes</a></li>
    <li><a class="dropdown-item" href="articles/gene_ontology.html">Gene Ontology in OmnipathR</a></li>
    <li><a class="dropdown-item" href="articles/nichenet.html">Using NicheNet with OmnipathR</a></li>
    <li><a class="dropdown-item" href="articles/omnipath_intro.html">OmnipathR: an R client for the OmniPath web service</a></li>
    <li><a class="dropdown-item" href="articles/paths.html">Pathway construction</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/saezlab/OmnipathR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/omnipathdb" aria-label="Go to the Twitter of OmniPath"><span class="fa fab fa-twitter fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://omnipathdb.org/" aria-label="Go to the OmniPath homepage"><span class="fa fas fa-university"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p><img src="reference/figures/logo_omnipath.png" align="right" height="140"></p>
<div class="section level1">
<div class="page-header">
<img src="logo.svg" class="logo" alt=""><h1 id="omnipathr">OmnipathR<a class="anchor" aria-label="anchor" href="#omnipathr"></a>
</h1>
</div>
<p>An <code>R</code> client for the <strong>OmniPath</strong> web service and many other resources.</p>
<div class="section level2">
<h2 id="package-contents">Package contents<a class="anchor" aria-label="anchor" href="#package-contents"></a>
</h2>
<ul>
<li>Client for the OmniPath web service</li>
<li>Functions for post-processing OmniPath data</li>
<li>Ensembl BioMart, UniProt and KEGG clients</li>
<li>Access to other databases (18+ resources, see below)</li>
<li>Integration to NicheNet, a method to infer ligand activities from transcriptomics data</li>
</ul>
<div class="section level3">
<h3 id="omnipath-web-service-client">OmniPath web service client<a class="anchor" aria-label="anchor" href="#omnipath-web-service-client"></a>
</h3>
<p><em>OmnipathR</em> retrieves the data from the OmniPath web service at</p>
<p><a href="https://omnipathdb.org/" class="external-link uri">https://omnipathdb.org/</a></p>
<p>The web service implements a very simple REST-style API. This package makes requests by the HTTP protocol to retrieve the data. Hence, fast Internet access is required for a proper use of <em>OmnipathR</em>.</p>
<div class="section level4">
<h4 id="what-is-omnipath">What is OmniPath?<a class="anchor" aria-label="anchor" href="#what-is-omnipath"></a>
</h4>
<p>OmniPath is a database of:</p>
<ul>
<li>Protein-protein, TF target and miRNA-mRNA interactions</li>
<li>Enzyme-PTM relationships</li>
<li>Protein complexes</li>
<li>Annotations of protein function, structure, localization, expression</li>
<li>Intercellular communication roles of proteins</li>
</ul>
<p>To learn more about OmniPath, you can visit its <a href="https://omnipathdb.org/" class="external-link">website</a>, or read our recent <a href="https://www.embopress.org/doi/full/10.15252/msb.20209923" class="external-link">publication</a> or our first <a href="https://www.nature.com/articles/nmeth.4077" class="external-link">paper from 2016</a>, especially its <a href="https://static-content.springer.com/esm/art%3A10.1038%2Fnmeth.4077/MediaObjects/41592_2016_BFnmeth4077_MOESM495_ESM.pdf" class="external-link">supplementary material</a>.</p>
</div>
</div>
<div class="section level3">
<h3 id="access-to-further-resources">Access to further resources<a class="anchor" aria-label="anchor" href="#access-to-further-resources"></a>
</h3>
<p>The package provides access to a number of other databases: BioPlex, ConsensusPathDB, EVEX, Gene Ontology, Guide to Pharmacology (IUPHAR/BPS), Harmonizome, HTRIdb, Human Phenotype Ontology, InWeb InBioMap, KEGG Pathway, Pathway Commons, PrePPI, Ramilowski et al. 2015, RegNetwork, ReMap, TF census, TRRUST and Vinayagam et al. 2011.</p>
</div>
</div>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<p>The latest version of the reference manual is available from <a href="https://static.omnipathdb.org/omnipathr_manual.pdf" class="external-link uri">https://static.omnipathdb.org/omnipathr_manual.pdf</a>. Tutorials can be found at <a href="https://workflows.omnipathdb.org/" class="external-link uri">https://workflows.omnipathdb.org/</a>. Scroll down for quick start examples.</p>
</div>
<div class="section level2">
<h2 id="omnipath-query-types">OmniPath query types<a class="anchor" aria-label="anchor" href="#omnipath-query-types"></a>
</h2>
<p>We provide here a brief summary of the data available through <em>OmnipathR</em>. <em>OmnipathR</em> provides access to 5 types of queries:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Interactions</strong>: protein-protein interactions from different datasets.</li>
<li>
<strong>Enzyme-substrate</strong>: enzyme-PTM (post-translational modification) relationships.</li>
<li>
<strong>Complexes</strong>: comprehensive database of more than 22000 protein complexes.</li>
<li>
<strong>Annotations</strong>: large variety of data about protein and complex features.</li>
<li>
<strong>Intercell</strong>: information on the roles in inter-cellular signaling.</li>
</ol>
<p>For more detailed information, we recommend you to visit the following sites:</p>
<p><a href="https://omnipathdb.org/" class="external-link uri">https://omnipathdb.org/</a></p>
<p><a href="https://omnipathdb.org/info" class="external-link uri">https://omnipathdb.org/info</a></p>
<p><a href="https://github.com/saezlab/pypath/blob/master/webservice.rst" class="external-link uri">https://github.com/saezlab/pypath/blob/master/webservice.rst</a></p>
<p><a href="https://saezlab.github.io/OmnipathR/articles/omnipath_intro.html" class="uri">https://saezlab.github.io/OmnipathR/articles/omnipath_intro.html</a></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>First of all, you need a current version of <code>R</code> (<a href="https://r-project.org" class="external-link uri">https://r-project.org</a>).</p>
<div class="section level3">
<h3 id="from-bioconductor">From Bioconductor<a class="anchor" aria-label="anchor" href="#from-bioconductor"></a>
</h3>
<p><em>OmnipathR</em> is a freely available package deposited on <em>Bioconductor</em> and <em>Github</em>: (<a href="https://bioconductor.org/" class="external-link uri">https://bioconductor.org/</a>, <a href="https://github.com/saezlab/OmnipathR" class="external-link uri">https://github.com/saezlab/OmnipathR</a>).</p>
<p>You can install it by running the following commands on a <code>R</code> console:</p>
<p>```{r} if (!requireNamespace(‘BiocManager’, quietly = TRUE)) install.packages(‘BiocManager’)</p>
</div>
</div>
<div class="section level2">
<h2 id="last-release-in-bioconductor">Last release in Bioconductor<a class="anchor" aria-label="anchor" href="#last-release-in-bioconductor"></a>
</h2>
<p>BiocManager::install(‘OmnipathR’, version = ‘3.12’) ## Development version with the latest updates BiocManager::install(‘OmnipathR’, version = ‘devel’)</p>
<pre><code>
### From github

We add new features to OmnipathR way more often than the Bioconductor release
frequency. To make use of the recent developments, you can use `devtools` to
install the package directly from github:

```{r}
require(devtools)
install_github('saezlab/OmnipathR')</code></pre>
</div>
<div class="section level2">
<h2 id="getting-started-and-some-usage-examples">Getting started and some usage examples<a class="anchor" aria-label="anchor" href="#getting-started-and-some-usage-examples"></a>
</h2>
<p>To get started, we strongly recommend to read our main vignette in order to deal with the different types of queries and handle the data they return:</p>
<p><a href="https://saezlab.github.io/OmnipathR/articles/omnipath_intro.html" class="uri">https://saezlab.github.io/OmnipathR/articles/omnipath_intro.html</a></p>
<p>You can also check the manual:</p>
<p><a href="https://saezlab.github.io/OmnipathR/reference/index.html" class="uri">https://saezlab.github.io/OmnipathR/reference/index.html</a></p>
<p>In addition, we provide here some examples for a quick start:</p>
<p><code>{r} library(OmnipathR)</code></p>
<p>Download human protein-protein interactions from the specified resources:</p>
<p><code>{r} interactions &lt;- omnipath(     resources = c('SignaLink3', 'PhosphoSite', 'SIGNOR') )</code></p>
<p>Download human enzyme-PTM relationships from the specified resources:</p>
<p><code>{r} enzsub &lt;- enzyme_substrate(resources = c('PhosphoSite', 'SIGNOR'))</code></p>
<p>Convert both data frames into networks (<code>igraph</code> objects) <code>{r} ptms_g = ptms_graph(ptms = enzsub) OPI_g = interaction_graph(interactions = interactions)</code></p>
<p>Print some interactions in a nice format: ```{r} print_interactions(head(interactions))</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>      source interaction         target n_resources n_references</span></code></pre></div>
<p>4 SRC (P12931) ==( + )==&gt; TRPV1 (Q8NER1) 9 6 2 PRKG1 (Q13976) ==( - )==&gt; TRPC6 (Q9Y210) 7 5 1 PRKG1 (Q13976) ==( - )==&gt; TRPC3 (Q13507) 9 2 5 LYN (P07948) ==( + )==&gt; TRPV4 (Q9HBA0) 9 2 6 PTPN1 (P18031) ==( - )==&gt; TRPV6 (Q9H1D0) 3 2 3 PRKACA (P17612) ==( + )==&gt; TRPV1 (Q8NER1) 6 1</p>
<pre><code>
Find interactions between a specific kinase and a specific substrate:
```{r}
print_interactions(dplyr::filter(enzsub,enzyme_genesymbol=='MAP2K1',
  substrate_genesymbol=='MAPK3'))

           enzyme interaction           substrate    modification n_resources
1 MAP2K1 (Q02750)       ====&gt; MAPK3_Y204 (P27361) phosphorylation           8
2 MAP2K1 (Q02750)       ====&gt; MAPK3_T202 (P27361) phosphorylation           8
3 MAP2K1 (Q02750)       ====&gt; MAPK3_Y210 (P27361) phosphorylation           2
4 MAP2K1 (Q02750)       ====&gt; MAPK3_T207 (P27361) phosphorylation           2
</code></pre>
<p>Find the shortest paths on the directed network between proteins: ```{r} print_path_es(shortest_paths(OPI_g,from = ‘TYRO3’,to = ‘STAT3’, output = ‘epath’)$epath[<a href="https://omnipathdb.org/" class="external-link">1</a>],OPI_g)</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>       source interaction          target n_resources n_references</span></code></pre></div>
<p>1 TYRO3 (Q06418) ==( ? )==&gt; AKT1 (P31749) 2 0 2 AKT1 (P31749) ==( - )==&gt; DAB2IP (Q5VWQ8) 3 1 3 DAB2IP (Q5VWQ8) ==( - )==&gt; STAT3 (P40763) 1 1</p>
<pre><code>
Find all shortest paths between proteins:
```{r}
print_path_vs(all_shortest_paths(OPI_g,from = 'DYRK2',to = 'MAPKAPK2')$res,OPI_g)
Pathway 1: DYRK2 -&gt; TBK1 -&gt; NFKB1 -&gt; MAP3K8 -&gt; MAPK3 -&gt; MAPKAPK2
Pathway 2: DYRK2 -&gt; TBK1 -&gt; AKT3 -&gt; MAP3K8 -&gt; MAPK3 -&gt; MAPKAPK2
Pathway 3: DYRK2 -&gt; TBK1 -&gt; AKT2 -&gt; MAP3K8 -&gt; MAPK3 -&gt; MAPKAPK2
Pathway 4: DYRK2 -&gt; TBK1 -&gt; AKT1 -&gt; MAP3K8 -&gt; MAPK3 -&gt; MAPKAPK2
Pathway 5: DYRK2 -&gt; TBK1 -&gt; AKT3 -&gt; PEA15 -&gt; MAPK3 -&gt; MAPKAPK2
Pathway 6: DYRK2 -&gt; TBK1 -&gt; AKT2 -&gt; PEA15 -&gt; MAPK3 -&gt; MAPKAPK2
.....</code></pre>
</div>
<div class="section level2">
<h2 id="alternatives">Alternatives<a class="anchor" aria-label="anchor" href="#alternatives"></a>
</h2>
<div class="section level3">
<h3 id="python">Python<a class="anchor" aria-label="anchor" href="#python"></a>
</h3>
<p>A similar web service client is available for Python:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&lt;</span>https<span class="sc">:</span><span class="er">//</span>github.com<span class="sc">/</span>saezlab<span class="sc">/</span>omnipath<span class="sc">&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cytoscape">Cytoscape<a class="anchor" aria-label="anchor" href="#cytoscape"></a>
</h3>
<p>The OmniPath Cytoscape app provides access to the interactions, enzyme-PTM relationships and some of the annotations:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="sc">&lt;</span>https<span class="sc">:</span><span class="er">//</span>apps.cytoscape.org<span class="sc">/</span>apps<span class="sc">/</span>omnipath<span class="sc">&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="customization">Customization<a class="anchor" aria-label="anchor" href="#customization"></a>
</h3>
<p>The <a href="https://github.com/saezlab/pypath" class="external-link"><em>pypath</em></a> Python module is a tool for building the OmniPath databases in a fully customizable way. We recommend to use pypath if you want to:</p>
<ul>
<li>Tailor the database building to your needs</li>
<li>Include resources not available in the public web service</li>
<li>Use the rich Python APIs available for the database objects</li>
<li>Make sure the data from the original sources is the most up-to-date</li>
<li>Use the methods in <code>pypath.inputs</code> to download data from resources</li>
<li>Use the various extra tools in <code>pypath.utils</code>, e.g. for identifier translation, homology translation, querying Gene Ontology, working with protein sequences, processing BioPAX, etc.</li>
</ul>
<p>With <em>pypath</em> it’s also possible to run your own web service and serve your custom databases to the <em>OmnipathR</em> R client and the <em>omnipath</em> Python client.</p>
</div>
</div>
<div class="section level2">
<h2 id="feedbacks-bug-reports-features">Feedbacks, bug reports, features<a class="anchor" aria-label="anchor" href="#feedbacks-bug-reports-features"></a>
</h2>
<p>Feedbacks and bug reports are always very welcome!</p>
<p>Please use the Github issue page to report bugs or for questions:</p>
<p><a href="https://github.com/saezlab/OmnipathR/issues" class="external-link uri">https://github.com/saezlab/OmnipathR/issues</a></p>
<p>Many thanks for using <em>OmnipathR</em>!</p>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/OmnipathR" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/saezlab/OmnipathR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/saezlab/OmnipathR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing OmnipathR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Alberto Valdeolivas <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-5482-9023" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Denes Turei <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0002-7249-9379" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Attila Gabor <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-0776-1182" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Diego Mananes <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-7247-6794" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Aurelien Dugourd <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-0714-028X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Valdeolivas, Denes Turei, Attila Gabor, Diego Mananes, Aurelien Dugourd.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
