<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="A client for the OmniPath web service
    (https://www.omnipathdb.org) and many other resources.
    It also includes functions to transform and pretty print
    some of the downloaded data, functions to access a number
    of other resources such as BioPlex, ConsensusPathDB, EVEX,
    Gene Ontology, Guide to Pharmacology (IUPHAR/BPS), Harmonizome,
    HTRIdb, Human Phenotype Ontology, InWeb InBioMap, KEGG Pathway,
    Pathway Commons, Ramilowski et al. 2015, RegNetwork, ReMap, TF
    census, TRRUST and Vinayagam et al. 2011. Furthermore, OmnipathR
    features a close integration  with the NicheNet method for
    ligand activity prediction from transcriptomics data, and its
    R implementation `nichenetr` (available only on github).">
<title>OmniPath web service client and more • OmnipathR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="deps/bs3compat-0.2.5.1/tabs.js"></script><script src="deps/bs3compat-0.2.5.1/bs3compat.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="OmniPath web service client and more">
<meta property="og:description" content="A client for the OmniPath web service
    (https://www.omnipathdb.org) and many other resources.
    It also includes functions to transform and pretty print
    some of the downloaded data, functions to access a number
    of other resources such as BioPlex, ConsensusPathDB, EVEX,
    Gene Ontology, Guide to Pharmacology (IUPHAR/BPS), Harmonizome,
    HTRIdb, Human Phenotype Ontology, InWeb InBioMap, KEGG Pathway,
    Pathway Commons, Ramilowski et al. 2015, RegNetwork, ReMap, TF
    census, TRRUST and Vinayagam et al. 2011. Furthermore, OmnipathR
    features a close integration  with the NicheNet method for
    ligand activity prediction from transcriptomics data, and its
    R implementation `nichenetr` (available only on github).">
<meta property="og:image" content="https://saezlab.github.io/OmnipathR/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-10"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-10');
</script>
</head>
<body data-spy="scroll" data-target="#toc" data-headroom>
    

      <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    <a class="navbar-brand mb-0 h1" href="index.html">OmnipathR</a>

    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <button type="button" id="version-badge" class="badge badge-default d-none d-lg-block" data-toggle="tooltip" data-placement="bottom" title="" data-original-title="Released version">3.1.2</button>

    <div id="navbar" class="collapse navbar-collapse">
      <ul class="navbar-nav mr-auto ml-3">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/bioc_workshop.html">OmniPath Bioconductor workshop</a>
    <a class="dropdown-item" href="articles/drug_targets.html">Building protein networks around drug-targets using OmnipathR</a>
    <a class="dropdown-item" href="articles/gene_ontology.html">Gene Ontology in OmnipathR</a>
    <a class="dropdown-item" href="articles/nichenet.html">Using NicheNet with OmnipathR</a>
    <a class="dropdown-item" href="articles/omnipath_intro.html">OmnipathR: an R client for the OmniPath web service</a>
    <a class="dropdown-item" href="articles/paths.html">Pathway construction</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control mr-sm-2" aria-label="Search" name="search-input" data-search-index="/search.json" id="search-input" placeholder="Search..." autocomplete="off">
</form>

      <ul class="navbar-nav ml-1">
<li class="nav-item">
  <a class="nav-link" href="https://github.com/saezlab/OmnipathR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="https://twitter.com/omnipathdb">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="https://omnipathdb.org/">
    <span class="fas fa fas fa-university"></span>
     
  </a>
</li>
      </ul>
</div>
  </div>
</nav><div class="container template-home">
<div class="row">
  <div class="contents col-md-9">

<img src="package-logo.png" class="pkg-logo" alt="" width="120"><p><img src="reference/figures/logo_omnipath.png" align="right" height="140"></p>
<div id="omnipathr" class="section level1">
<div class="pb-2 mt-4 mb-2 border-bottom"><h1 class="hasAnchor">
<a href="#omnipathr" class="anchor" aria-hidden="true"></a>OmnipathR</h1></div>
<p>An <code>R</code> client for the <strong>OmniPath</strong> web service and many other resources.</p>
<div id="package-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#package-contents" class="anchor" aria-hidden="true"></a>Package contents</h2>
<ul>
<li>Client for the OmniPath web service</li>
<li>Functions for post-processing OmniPath data</li>
<li>Access to other databases (18+ resources, see below)</li>
<li>Integration to NicheNet, a method to infer ligand activities from transcriptomics data</li>
</ul>
<div id="omnipath-web-service-client" class="section level3">
<h3 class="hasAnchor">
<a href="#omnipath-web-service-client" class="anchor" aria-hidden="true"></a>OmniPath web service client</h3>
<p><em>OmnipathR</em> retrieves the data from the OmniPath web service at</p>
<p><a href="https://omnipathdb.org/" class="external-link uri">https://omnipathdb.org/</a></p>
<p>The web service implements a very simple REST style API. This package make requests by the HTTP protocol to retreive the data. Hence, fast Internet access is required for a proper use of <em>OmnipathR</em>.</p>
<div id="what-is-omnipath" class="section level4">
<h4 class="hasAnchor">
<a href="#what-is-omnipath" class="anchor" aria-hidden="true"></a>What is OmniPath?</h4>
<p>OmniPath is a database of:</p>
<ul>
<li>Protein-protein, TF target and miRNA-mRNA interactions</li>
<li>Enzyme-PTM relationships</li>
<li>Protein complexes</li>
<li>Annotations of protein function, structure, localization, expression</li>
<li>Intercellular communication roles of proteins</li>
</ul>
<p>To learn more about OmniPath, you can visit its <a href="https://omnipathdb.org/" class="external-link">website</a>, or read our recent <a href="https://www.embopress.org/doi/full/10.15252/msb.20209923" class="external-link">publication</a> or our first <a href="https://www.nature.com/articles/nmeth.4077" class="external-link">paper from 2016</a>, especially its <a href="https://static-content.springer.com/esm/art%3A10.1038%2Fnmeth.4077/MediaObjects/41592_2016_BFnmeth4077_MOESM495_ESM.pdf" class="external-link">supplementary material</a>.</p>
</div>
</div>
<div id="access-to-further-resources" class="section level3">
<h3 class="hasAnchor">
<a href="#access-to-further-resources" class="anchor" aria-hidden="true"></a>Access to further resources</h3>
<p>The package provides access to a number of other databases: BioPlex, ConsensusPathDB, EVEX, Gene Ontology, Guide to Pharmacology (IUPHAR/BPS), Harmonizome, HTRIdb, Human Phenotype Ontology, InWeb InBioMap, KEGG Pathway, Pathway Commons, PrePPI, Ramilowski et al. 2015, RegNetwork, ReMap, TF census, TRRUST and Vinayagam et al. 2011.</p>
</div>
</div>
<div id="documentation" class="section level2">
<h2 class="hasAnchor">
<a href="#documentation" class="anchor" aria-hidden="true"></a>Documentation</h2>
<p>The latest version of the reference manual is available from <a href="https://static.omnipathdb.org/omnipathr_manual.pdf" class="external-link uri">https://static.omnipathdb.org/omnipathr_manual.pdf</a>. Tutorials can be found at <a href="https://workflows.omnipathdb.org/" class="external-link uri">https://workflows.omnipathdb.org/</a>. Sroll down for quick start examples.</p>
</div>
<div id="omnipath-query-types" class="section level2">
<h2 class="hasAnchor">
<a href="#omnipath-query-types" class="anchor" aria-hidden="true"></a>OmniPath query types</h2>
<p>We provide here a brief summary about the data available through <em>OmnipathR</em>. <em>OmnipathR</em> provides access to 5 types of queries:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Interactions</strong>: protein-protein interactions from different datasets.</li>
<li>
<strong>Enzyme-substrate</strong>: enzyme-PTM (post-translational modification) relationships.</li>
<li>
<strong>Complexes</strong>: comprehensive database of more than 22000 protein complexes.</li>
<li>
<strong>Annotations</strong>: large variety of data about proteins and complexes features.</li>
<li>
<strong>Intercell</strong>: information on the roles in inter-cellular signaling.</li>
</ol>
<p>For a more detailed information, we recommend you to visit the following sites:</p>
<p><a href="https://omnipathdb.org/" class="external-link uri">https://omnipathdb.org/</a></p>
<p><a href="https://omnipathdb.org/info" class="external-link uri">https://omnipathdb.org/info</a></p>
<p><a href="https://github.com/saezlab/pypath/blob/master/webservice.rst" class="external-link uri">https://github.com/saezlab/pypath/blob/master/webservice.rst</a></p>
<p><a href="https://saezlab.github.io/OmnipathR/articles/omnipath_intro.html" class="uri">https://saezlab.github.io/OmnipathR/articles/omnipath_intro.html</a></p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor" aria-hidden="true"></a>Installation</h2>
<p>First of all, you need a current version of <code>R</code> (<a href="https://r-project.org" class="external-link uri">https://r-project.org</a>).</p>
<div id="from-bioconductor" class="section level3">
<h3 class="hasAnchor">
<a href="#from-bioconductor" class="anchor" aria-hidden="true"></a>From Bioconductor</h3>
<p><em>OmnipathR</em> is a freely available package deposited on <em>Bioconductor</em> and <em>Github</em>: (<a href="https://bioconductor.org/" class="external-link uri">https://bioconductor.org/</a>, <a href="https://github.com/saezlab/OmnipathR" class="external-link uri">https://github.com/saezlab/OmnipathR</a>).</p>
<p>You can install it by running the following commands on a <code>R</code> console:</p>
<pre class="{r}"><code>if (!requireNamespace('BiocManager', quietly = TRUE))
    install.packages('BiocManager')

## Last release in Bioconductor
BiocManager::install('OmnipathR', version = '3.12')
## Development version with the lastest updates
BiocManager::install('OmnipathR', version = 'devel')</code></pre>
</div>
<div id="from-github" class="section level3">
<h3 class="hasAnchor">
<a href="#from-github" class="anchor" aria-hidden="true"></a>From github</h3>
<p>We add new features to OmnipathR way more often than the Bioconductor release frequency. To make use of the recent developments, you can use <code>devtools</code> to install the package directly from github:</p>
<pre class="{r}"><code><a href="https://devtools.r-lib.org/" class="external-link">require(devtools)
install_github('saezlab/OmnipathR')</a></code></pre>
</div>
</div>
<div id="getting-started-and-some-usage-examples" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started-and-some-usage-examples" class="anchor" aria-hidden="true"></a>Getting started and some usage examples</h2>
<p>To get started, we strongly recommend to read our main vignette in order to deal with the different types of queries and handle the data they return:</p>
<p><a href="https://saezlab.github.io/OmnipathR/articles/omnipath_intro.html" class="uri">https://saezlab.github.io/OmnipathR/articles/omnipath_intro.html</a></p>
<p>You can also check the manual:</p>
<p><a href="https://saezlab.github.io/OmnipathR/reference/index.html" class="uri">https://saezlab.github.io/OmnipathR/reference/index.html</a></p>
<p>In addition, we provide here some examples for a quick start:</p>
<pre class="{r}"><code><a href="https://saezlab.github.io/OmnipathR/">library(OmnipathR)</a></code></pre>
<p>Download human protein-protein interactions from the specified resources:</p>
<pre class="{r}"><code>interactions &lt;- import_omnipath_interactions(
    resources = c('SignaLink3', 'PhosphoSite', 'SIGNOR')
)</code></pre>
<p>Download human enzyme-PTM relationships from the specified resources:</p>
<pre class="{r}"><code>enzsub &lt;- import_omnipath_enzsub(resources = c('PhosphoSite', 'SIGNOR'))</code></pre>
<p>Convert both data frames into networks (<code>igraph</code> objects)</p>
<pre class="{r}"><code>ptms_g = ptms_graph(ptms = enzsub)
OPI_g = interaction_graph(interactions = interactions)</code></pre>
<p>Print some interactions in a nice format:</p>
<pre class="{r}"><code>print_interactions(head(interactions))

          source interaction         target n_resources n_references
4    SRC (P12931)  ==( + )==&gt; TRPV1 (Q8NER1)           9            6
2  PRKG1 (Q13976)  ==( - )==&gt; TRPC6 (Q9Y210)           7            5
1  PRKG1 (Q13976)  ==( - )==&gt; TRPC3 (Q13507)           9            2
5    LYN (P07948)  ==( + )==&gt; TRPV4 (Q9HBA0)           9            2
6  PTPN1 (P18031)  ==( - )==&gt; TRPV6 (Q9H1D0)           3            2
3 PRKACA (P17612)  ==( + )==&gt; TRPV1 (Q8NER1)           6            1</code></pre>
<p>Find interactions between a specific kinase and a specific substrate:</p>
<pre class="{r}"><code>print_interactions(dplyr::filter(enzsub,enzyme_genesymbol=='MAP2K1',
  substrate_genesymbol=='MAPK3'))

           enzyme interaction           substrate    modification n_resources
1 MAP2K1 (Q02750)       ====&gt; MAPK3_Y204 (P27361) phosphorylation           8
2 MAP2K1 (Q02750)       ====&gt; MAPK3_T202 (P27361) phosphorylation           8
3 MAP2K1 (Q02750)       ====&gt; MAPK3_Y210 (P27361) phosphorylation           2
4 MAP2K1 (Q02750)       ====&gt; MAPK3_T207 (P27361) phosphorylation           2
</code></pre>
<p>Find shortest paths on the directed network between proteins:</p>
<pre class="{r}"><code>print_path_es(shortest_paths(OPI_g,from = 'TYRO3',to = 'STAT3',
    output = 'epath')$epath[[1]],OPI_g)

           source interaction          target n_resources n_references
1  TYRO3 (Q06418)  ==( ? )==&gt;   AKT1 (P31749)           2            0
2   AKT1 (P31749)  ==( - )==&gt; DAB2IP (Q5VWQ8)           3            1
3 DAB2IP (Q5VWQ8)  ==( - )==&gt;  STAT3 (P40763)           1            1</code></pre>
<p>Find all shortest paths between proteins:</p>
<pre class="{r}"><code>print_path_vs(all_shortest_paths(OPI_g,from = 'DYRK2',to = 'MAPKAPK2')$res,OPI_g)
Pathway 1: DYRK2 -&gt; TBK1 -&gt; NFKB1 -&gt; MAP3K8 -&gt; MAPK3 -&gt; MAPKAPK2
Pathway 2: DYRK2 -&gt; TBK1 -&gt; AKT3 -&gt; MAP3K8 -&gt; MAPK3 -&gt; MAPKAPK2
Pathway 3: DYRK2 -&gt; TBK1 -&gt; AKT2 -&gt; MAP3K8 -&gt; MAPK3 -&gt; MAPKAPK2
Pathway 4: DYRK2 -&gt; TBK1 -&gt; AKT1 -&gt; MAP3K8 -&gt; MAPK3 -&gt; MAPKAPK2
Pathway 5: DYRK2 -&gt; TBK1 -&gt; AKT3 -&gt; PEA15 -&gt; MAPK3 -&gt; MAPKAPK2
Pathway 6: DYRK2 -&gt; TBK1 -&gt; AKT2 -&gt; PEA15 -&gt; MAPK3 -&gt; MAPKAPK2
.....</code></pre>
</div>
<div id="alternatives" class="section level2">
<h2 class="hasAnchor">
<a href="#alternatives" class="anchor" aria-hidden="true"></a>Alternatives</h2>
<div id="python" class="section level3">
<h3 class="hasAnchor">
<a href="#python" class="anchor" aria-hidden="true"></a>Python</h3>
<p>A similar web service client is available for Python:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&lt;</span>https<span class="sc">:</span><span class="er">//</span>github.com<span class="sc">/</span>saezlab<span class="sc">/</span>omnipath<span class="sc">&gt;</span></span></code></pre></div>
</div>
<div id="cytoscape" class="section level3">
<h3 class="hasAnchor">
<a href="#cytoscape" class="anchor" aria-hidden="true"></a>Cytoscape</h3>
<p>The OmniPath Cytoscape app provides access to the interactions, enzyme-PTM relationships and some of the annotations:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&lt;</span>https<span class="sc">:</span><span class="er">//</span>apps.cytoscape.org<span class="sc">/</span>apps<span class="sc">/</span>omnipath<span class="sc">&gt;</span></span></code></pre></div>
</div>
<div id="customization" class="section level3">
<h3 class="hasAnchor">
<a href="#customization" class="anchor" aria-hidden="true"></a>Customization</h3>
<p>The <a href="https://github.com/saezlab/pypath" class="external-link"><em>pypath</em></a> Python module is a tool for building the OmniPath databases in a fully customizable way. We recommend to use pypath if you want to:</p>
<ul>
<li>Tailor the database building to your needs</li>
<li>Include resources not available in the public web service</li>
<li>Use the rich Python APIs available for the database objects</li>
<li>Make sure the data from the original sources is the most up-to-date</li>
<li>Use the methods in <code>pypath.inputs</code> to download data from resources</li>
<li>Use the various extra tools in <code>pypath.utils</code>, e.g. for identifier translation, homology translation, querying Gene Ontology, working with protein sequences, processing BioPAX, etc.</li>
</ul>
<p>With <em>pypath</em> it’s also possible to run your own web service and serve your custom databases to the <em>OmnipathR</em> R client and the <em>omnipath</em> Python cient.</p>
</div>
</div>
<div id="feedbacks-bug-reports-features" class="section level2">
<h2 class="hasAnchor">
<a href="#feedbacks-bug-reports-features" class="anchor" aria-hidden="true"></a>Feedbacks, bug reports, features</h2>
<p>Feedbacks and bugreports are always very welcome!</p>
<p>Please use the Github issue page to report bugs or for questions:</p>
<p><a href="https://github.com/saezlab/OmnipathR/issues" class="external-link uri">https://github.com/saezlab/OmnipathR/issues</a></p>
<p>Many thanks for using <em>OmnipathR</em>!</p>
</div>
</div>

  </div>

  <div class="col-md-3 d-none d-md-block" id="pkgdown-sidebar">
      <div class="sticky-top">
        <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/OmnipathR">https://​www.bioconductor.org/​packages/​OmnipathR</a>
</li>
<li>Browse source code at <br><a href="https://github.com/saezlab/OmnipathR/">https://​github.com/​saezlab/​OmnipathR/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/saezlab/OmnipathR/issues">https://​github.com/​saezlab/​OmnipathR/​issues</a>
</li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing OmnipathR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Alberto Valdeolivas <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-5482-9023" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Denes Turei <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0002-7249-9379" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Attila Gabor <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-0776-1182" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



      </div>
  </div>
</div>


      <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Alberto Valdeolivas, Denes Turei, Attila Gabor.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  

  

  </body>
</html>
