<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Prior knowledge network (PKN) for COSMOS — cosmos_pkn • OmnipathR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Prior knowledge network (PKN) for COSMOS — cosmos_pkn"><meta name="description" content="The prior knowledge network (PKN) used by COSMOS is a network of
heterogenous causal interactions: it contains protein-protein,
reactant-enzyme and enzyme-product interactions. It is a combination of
multiple resources:
Genome-scale metabolic model (GEM) from Chalmers Sysbio (Wang et
    al., 2021.)
Network of chemical-protein interactions from STITCH
    (https://stitch.embl.de/)
Protein-protein interactions from Omnipath (Türei
    et al., 2021)
This function downloads, processes and combines the resources above. With
all downloads and processing the build might take 30-40 minutes. Data is
cached at various levels of processing, shortening processing times. With
all data downloaded and HMDB ID translation data preprocessed, the build
takes 3-4 minutes; the complete PKN is also saved in the cache, if this is
available, loading it takes only a few seconds."><meta property="og:description" content="The prior knowledge network (PKN) used by COSMOS is a network of
heterogenous causal interactions: it contains protein-protein,
reactant-enzyme and enzyme-product interactions. It is a combination of
multiple resources:
Genome-scale metabolic model (GEM) from Chalmers Sysbio (Wang et
    al., 2021.)
Network of chemical-protein interactions from STITCH
    (https://stitch.embl.de/)
Protein-protein interactions from Omnipath (Türei
    et al., 2021)
This function downloads, processes and combines the resources above. With
all downloads and processing the build might take 30-40 minutes. Data is
cached at various levels of processing, shortening processing times. With
all data downloaded and HMDB ID translation data preprocessed, the build
takes 3-4 minutes; the complete PKN is also saved in the cache, if this is
available, loading it takes only a few seconds."><meta property="og:image" content="https://saezlab.github.io/OmnipathR/logo.svg"><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-10"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-10');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">OmnipathR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.13.25</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bioc_workshop.html">OmniPath Bioconductor workshop</a></li>
    <li><a class="dropdown-item" href="../articles/cosmos.html">Building prior knowledge network (PKN) for COSMOS</a></li>
    <li><a class="dropdown-item" href="../articles/db_manager.html">The database manager in OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/drug_targets.html">Building protein networks around drug-targets using OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/extra_attrs.html">Resource specific interaction attributes</a></li>
    <li><a class="dropdown-item" href="../articles/gene_ontology.html">Gene Ontology in OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/nichenet.html">Using NicheNet with OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/omnipath_intro.html">OmnipathR: an R client for the OmniPath web service</a></li>
    <li><a class="dropdown-item" href="../articles/paths.html">Pathway construction</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/saezlab/OmnipathR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/omnipathdb" aria-label="Go to the Twitter of OmniPath"><span class="fa fab fa-twitter fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://omnipathdb.org/" aria-label="Go to the OmniPath homepage"><span class="fa fas fa-university"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Prior knowledge network (PKN) for COSMOS</h1>
      <small class="dont-index">Source: <a href="https://github.com/saezlab/OmnipathR/blob/HEAD/R/cosmos.R" class="external-link"><code>R/cosmos.R</code></a></small>
      <div class="d-none name"><code>cosmos_pkn.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The prior knowledge network (PKN) used by COSMOS is a network of
heterogenous causal interactions: it contains protein-protein,
reactant-enzyme and enzyme-product interactions. It is a combination of
multiple resources:</p><ul><li><p>Genome-scale metabolic model (GEM) from Chalmers Sysbio (Wang et
    al., 2021.)</p></li>
<li><p>Network of chemical-protein interactions from STITCH
    (<a href="https://stitch.embl.de/" class="external-link">https://stitch.embl.de/</a>)</p></li>
<li><p>Protein-protein interactions from Omnipath (Türei
    et al., 2021)</p></li>
</ul><p>This function downloads, processes and combines the resources above. With
all downloads and processing the build might take 30-40 minutes. Data is
cached at various levels of processing, shortening processing times. With
all data downloaded and HMDB ID translation data preprocessed, the build
takes 3-4 minutes; the complete PKN is also saved in the cache, if this is
available, loading it takes only a few seconds.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">cosmos_pkn</span><span class="op">(</span></span>
<span>  organism <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>  protein_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"uniprot"</span>, <span class="st">"genesymbol"</span><span class="op">)</span>,</span>
<span>  metabolite_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hmdb"</span>, <span class="st">"kegg"</span><span class="op">)</span>,</span>
<span>  chalmers_gem_metab_max_degree <span class="op">=</span> <span class="fl">400L</span>,</span>
<span>  stitch_score <span class="op">=</span> <span class="fl">700L</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-organism">organism<a class="anchor" aria-label="anchor" href="#arg-organism"></a></dt>
<dd><p>Character or integer: name or NCBI Taxonomy ID of an
organism. Supported organisms vary by resource: the Chalmers GEM is
available only for human, mouse, rat, fish, fly and worm. OmniPath can
be translated by orthology, but for non-vertebrate or less researched
taxa very few orthologues are available. STITCH is available for a
large number of organisms, please refer to their web page:
<a href="https://stitch.embl.de/" class="external-link">https://stitch.embl.de/</a>.</p></dd>


<dt id="arg-protein-ids">protein_ids<a class="anchor" aria-label="anchor" href="#arg-protein-ids"></a></dt>
<dd><p>Character: translate the protein identifiers to these ID
types. Each ID type results two extra columns in the output, for the
"source" and "target" sides of the interaction, respectively. The
default ID type for proteins depends on the resource, hence the "source"
and "target" columns are heterogenous. By default UniProt IDs and Gene
Symbols are included. The Gene Symbols used in the COSMOS PKN are
provided by Ensembl, and do not completely agree with the ones provided
by UniProt and used in OmniPath data by default.</p></dd>


<dt id="arg-metabolite-ids">metabolite_ids<a class="anchor" aria-label="anchor" href="#arg-metabolite-ids"></a></dt>
<dd><p>Character: translate the metabolite identifiers to
these ID types. Each ID type results two extra columns in the output,
for the "source" and "target" sides of the interaction, respectively.
The default ID type for metabolites depends on the resource, hence the
"source" and "target" columns are heterogenous. By default HMDB IDs and
KEGG IDs are included.</p></dd>


<dt id="arg-chalmers-gem-metab-max-degree">chalmers_gem_metab_max_degree<a class="anchor" aria-label="anchor" href="#arg-chalmers-gem-metab-max-degree"></a></dt>
<dd><p>Numeric: remove metabolites from the
Chalmers GEM network with defgrees larger than this. Useful to remove
cofactors and over-promiscuous metabolites.</p></dd>


<dt id="arg-stitch-score">stitch_score<a class="anchor" aria-label="anchor" href="#arg-stitch-score"></a></dt>
<dd><p>Include interactions from STITCH with combined
confidence score larger than this.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further parameters to <code><a href="omnipath-interactions.html">omnipath_interactions</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame of binary causal interations with effect signs,
    resource specific attributes and translated to the desired identifiers.
    The “record_id“ column identifies the original records within each
    resource. If one “record_id“ yields multiple records in the final data
    frame, it is the result of one-to-many ID translation or other
    processing steps. Before use, it is recommended to select one pair of ID
    type columns (by combining the preferred ones) and perform “distinct“
    by the identifier columns and sign.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Wang H, Robinson JL, Kocabas P, Gustafsson J, Anton M,
    Cholley PE, et al. Genome-scale metabolic network reconstruction of model
    animals as a platform for translational research. Proceedings of the
    National Academy of Sciences. 2021 Jul 27;118(30):e2102344118.</p>
<p>Türei D, Valdeolivas A, Gul L, Palacio‐Escat N, Klein M, Ivanova O, et al.
    Integrated intra‐ and intercellular signaling knowledge for multicellular
    omics analysis. Molecular Systems Biology. 2021 Mar;17(3):e9923.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index">
<ul><li><p><code><a href="chalmers_gem_network.html">chalmers_gem_network</a></code></p></li>
<li><p><code><a href="stitch_network.html">stitch_network</a></code></p></li>
<li><p><code><a href="omnipath_for_cosmos.html">omnipath_for_cosmos</a></code></p></li>
<li><p><code><a href="omnipath-interactions.html">omnipath-interactions</a></code></p></li>
</ul></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span>    <span class="va">human_cosmos</span> <span class="op">&lt;-</span> <span class="fu">cosmos_pkn</span><span class="op">(</span>organism <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Valdeolivas, Denes Turei, Attila Gabor, Diego Mananes, Aurelien Dugourd.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

