<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ OmnipathR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"><meta property="og:image" content="https://saezlab.github.io/OmnipathR/logo.svg"><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-10"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-10');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">OmnipathR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.15.12</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bioc_workshop.html">OmniPath Bioconductor workshop</a></li>
    <li><a class="dropdown-item" href="../articles/cosmos.html">Building prior knowledge network (PKN) for COSMOS</a></li>
    <li><a class="dropdown-item" href="../articles/db_manager.html">The database manager in OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/drug_targets.html">Building protein networks around drug-targets using OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/extra_attrs.html">Resource specific interaction attributes</a></li>
    <li><a class="dropdown-item" href="../articles/gene_ontology.html">Gene Ontology in OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/id_translation.html">Identifier translation</a></li>
    <li><a class="dropdown-item" href="../articles/nichenet.html">Using NicheNet with OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/omnipath_intro.html">OmnipathR: an R client for the OmniPath web service</a></li>
    <li><a class="dropdown-item" href="../articles/paths.html">Pathway construction</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/saezlab/OmnipathR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/omnipathdb" aria-label="Go to the Twitter of OmniPath"><span class="fa fab fa-twitter fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://omnipathdb.org/" aria-label="Go to the OmniPath homepage"><span class="fa fas fa-university"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="OmnipathR.html">OmnipathR-package</a></code> <code><a href="OmnipathR.html">OmnipathR</a></code>

        </dt>
        <dd>The OmnipathR package</dd>
      </dl><dl><dt>

          <code><a href="all_uniprot_acs.html">all_uniprot_acs()</a></code>

        </dt>
        <dd>All UniProt ACs for one organism</dd>
      </dl><dl><dt>

          <code><a href="all_uniprots.html">all_uniprots()</a></code>

        </dt>
        <dd>A table with all UniProt records</dd>
      </dl><dl><dt>

          <code><a href="ambiguity.html">ambiguity()</a></code>

        </dt>
        <dd>Inspect the ambiguity of a mapping</dd>
      </dl><dl><dt>

          <code><a href="ancestors.html">ancestors()</a></code>

        </dt>
        <dd>All ancestors in the ontology tree</dd>
      </dl><dl><dt>

          <code><a href="annotated_network.html">annotated_network()</a></code>

        </dt>
        <dd>Network interactions with annotations</dd>
      </dl><dl><dt>

          <code><a href="annotation_categories.html">annotation_categories()</a></code>

        </dt>
        <dd>Annotation categories and resources</dd>
      </dl><dl><dt>

          <code><a href="annotation_resources.html">annotation_resources()</a></code>

        </dt>
        <dd>Retrieves a list of available resources in the annotations database of OmniPath</dd>
      </dl><dl><dt>

          <code><a href="annotations.html">annotations()</a></code>

        </dt>
        <dd>Protein and gene annotations from OmniPath</dd>
      </dl><dl><dt>

          <code><a href="biomart_query.html">biomart_query()</a></code>

        </dt>
        <dd>Query the Ensembl BioMart web service</dd>
      </dl><dl><dt>

          <code><a href="bioplex1.html">bioplex1()</a></code>

        </dt>
        <dd>Downloads the BioPlex version 1.0 interaction dataset</dd>
      </dl><dl><dt>

          <code><a href="bioplex2.html">bioplex2()</a></code>

        </dt>
        <dd>Downloads the BioPlex version 2.0 interaction dataset</dd>
      </dl><dl><dt>

          <code><a href="bioplex3.html">bioplex3()</a></code>

        </dt>
        <dd>Downloads the BioPlex version 3.0 interaction dataset</dd>
      </dl><dl><dt>

          <code><a href="bioplex_all.html">bioplex_all()</a></code>

        </dt>
        <dd>Downloads all BioPlex interaction datasets</dd>
      </dl><dl><dt>

          <code><a href="bioplex_hct116_1.html">bioplex_hct116_1()</a></code>

        </dt>
        <dd>Downloads the BioPlex HCT116 version 1.0 interaction dataset</dd>
      </dl><dl><dt>

          <code><a href="bma_motif_es.html">bma_motif_es()</a></code>

        </dt>
        <dd>BMA motifs from a sequence of edges</dd>
      </dl><dl><dt>

          <code><a href="bma_motif_vs.html">bma_motif_vs()</a></code>

        </dt>
        <dd>Prints a BMA motif to the screen from a sequence of nodes, which can be copy/pasted into the BMA canvas</dd>
      </dl><dl><dt>

          <code><a href="chalmers_gem.html">chalmers_gem()</a></code>

        </dt>
        <dd>Genome scale metabolic model by Wang et al. 2021</dd>
      </dl><dl><dt>

          <code><a href="chalmers_gem_id_mapping_table.html">chalmers_gem_id_mapping_table()</a></code>

        </dt>
        <dd>Metabolite ID translation tables from Chalmers Sysbio</dd>
      </dl><dl><dt>

          <code><a href="chalmers_gem_id_type.html">chalmers_gem_id_type()</a></code>

        </dt>
        <dd>Metabolite identifier type label used in Chalmers Sysbio GEM</dd>
      </dl><dl><dt>

          <code><a href="chalmers_gem_metabolites.html">chalmers_gem_metabolites()</a></code>

        </dt>
        <dd>Metabolites from the Chalmers SysBio GEM (Wang et al., 2021)</dd>
      </dl><dl><dt>

          <code><a href="chalmers_gem_network.html">chalmers_gem_network()</a></code>

        </dt>
        <dd>Chalmers SysBio GEM in the form of gene-metabolite interactions</dd>
      </dl><dl><dt>

          <code><a href="chalmers_gem_raw.html">chalmers_gem_raw()</a></code>

        </dt>
        <dd>GEM matlab file from Chalmers Sysbio (Wang et al., 2021)</dd>
      </dl><dl><dt>

          <code><a href="chalmers_gem_reactions.html">chalmers_gem_reactions()</a></code>

        </dt>
        <dd>Reactions from the Chalmers SysBio GEM (Wang et al., 2021)</dd>
      </dl><dl><dt>

          <code><a href="common_name.html">common_name()</a></code>

        </dt>
        <dd>Common (English) names of organisms</dd>
      </dl><dl><dt>

          <code><a href="complex_genes.html">complex_genes()</a></code>

        </dt>
        <dd>Get all the molecular complexes for a given gene(s)</dd>
      </dl><dl><dt>

          <code><a href="complex_resources.html">complex_resources()</a></code>

        </dt>
        <dd>Retrieve a list of complex resources available in Omnipath</dd>
      </dl><dl><dt>

          <code><a href="complexes.html">complexes()</a></code>

        </dt>
        <dd>Protein complexes from OmniPath</dd>
      </dl><dl><dt>

          <code><a href="consensuspathdb_download.html">consensuspathdb_download()</a></code>

        </dt>
        <dd>Retrieves the ConsensusPathDB network</dd>
      </dl><dl><dt>

          <code><a href="consensuspathdb_raw_table.html">consensuspathdb_raw_table()</a></code>

        </dt>
        <dd>Downloads interaction data from ConsensusPathDB</dd>
      </dl><dl><dt>

          <code><a href="cosmos_pkn.html">cosmos_pkn()</a></code>

        </dt>
        <dd>Prior knowledge network (PKN) for COSMOS</dd>
      </dl><dl><dt>

          <code><a href="curated_ligand_receptor_interactions.html">curated_ligand_receptor_interactions()</a></code>

        </dt>
        <dd>Curated ligand-receptor interactions</dd>
      </dl><dl><dt>

          <code><a href="curated_ligrec_stats.html">curated_ligrec_stats()</a></code>

        </dt>
        <dd>Statistics about literature curated ligand-receptor interactions</dd>
      </dl><dl><dt>

          <code><a href="database_summary.html">database_summary()</a></code>

        </dt>
        <dd>Summary of the annotations and intercell database contents</dd>
      </dl><dl><dt>

          <code><a href="datasets_one_column.html">datasets_one_column()</a></code>

        </dt>
        <dd>Create a column with dataset names listed</dd>
      </dl><dl><dt>

          <code><a href="descendants.html">descendants()</a></code>

        </dt>
        <dd>All descendants in the ontology tree</dd>
      </dl><dl><dt>

          <code><a href="dot-omnipathr_options_defaults.html">.omnipathr_options_defaults</a></code>

        </dt>
        <dd>Default values for the package options</dd>
      </dl><dl><dt>

          <code><a href="ensembl_dataset.html">ensembl_dataset()</a></code>

        </dt>
        <dd>Ensembl dataset name from organism</dd>
      </dl><dl><dt>

          <code><a href="ensembl_id_mapping_table.html">ensembl_id_mapping_table()</a></code>

        </dt>
        <dd>Identifier translation table from Ensembl</dd>
      </dl><dl><dt>

          <code><a href="ensembl_id_type.html">ensembl_id_type()</a></code>

        </dt>
        <dd>Ensembl identifier type label</dd>
      </dl><dl><dt>

          <code><a href="ensembl_name.html">ensembl_name()</a></code>

        </dt>
        <dd>Ensembl identifiers of organisms</dd>
      </dl><dl><dt>

          <code><a href="ensembl_organisms.html">ensembl_organisms()</a></code>

        </dt>
        <dd>Organism names and identifiers from Ensembl</dd>
      </dl><dl><dt>

          <code><a href="ensembl_organisms_raw.html">ensembl_organisms_raw()</a></code>

        </dt>
        <dd>Table of Ensembl organisms</dd>
      </dl><dl><dt>

          <code><a href="ensembl_orthology.html">ensembl_orthology()</a></code>

        </dt>
        <dd>Orthologous gene pairs from Ensembl</dd>
      </dl><dl><dt>

          <code><a href="ensure_igraph.html">ensure_igraph()</a></code>

        </dt>
        <dd>Converts a network to igraph object unless it is already one</dd>
      </dl><dl><dt>

          <code><a href="enzsub_graph.html">enzsub_graph()</a></code>

        </dt>
        <dd>Enzyme-substrate graph</dd>
      </dl><dl><dt>

          <code><a href="enzsub_resources.html">enzsub_resources()</a></code>

        </dt>
        <dd>Retrieves a list of enzyme-substrate resources available in OmniPath</dd>
      </dl><dl><dt>

          <code><a href="enzyme_substrate.html">enzyme_substrate()</a></code>

        </dt>
        <dd>Enzyme-substrate (PTM) relationships from OmniPath</dd>
      </dl><dl><dt>

          <code><a href="evex_download.html">evex_download()</a></code>

        </dt>
        <dd>Interactions from the EVEX database</dd>
      </dl><dl><dt>

          <code><a href="evidences.html">evidences()</a></code>

        </dt>
        <dd>Show evidences for an interaction</dd>
      </dl><dl><dt>

          <code><a href="extra_attr_values.html">extra_attr_values()</a></code>

        </dt>
        <dd>Possible values of an extra attribute</dd>
      </dl><dl><dt>

          <code><a href="extra_attrs.html">extra_attrs()</a></code>

        </dt>
        <dd>Extra attribute names in an interaction data frame</dd>
      </dl><dl><dt>

          <code><a href="extra_attrs_to_cols.html">extra_attrs_to_cols()</a></code>

        </dt>
        <dd>New columns from extra attributes</dd>
      </dl><dl><dt>

          <code><a href="filter_by_resource.html">filter_by_resource()</a></code>

        </dt>
        <dd>Filters OmniPath data by resources</dd>
      </dl><dl><dt>

          <code><a href="filter_evidences.html">filter_evidences()</a></code>

        </dt>
        <dd>Filter evidences by dataset, resource and license</dd>
      </dl><dl><dt>

          <code><a href="filter_extra_attrs.html">filter_extra_attrs()</a></code>

        </dt>
        <dd>Filter interactions by extra attribute values</dd>
      </dl><dl><dt>

          <code><a href="filter_intercell.html">filter_intercell()</a></code>

        </dt>
        <dd>Filter intercell annotations</dd>
      </dl><dl><dt>

          <code><a href="filter_intercell_network.html">filter_intercell_network()</a></code>

        </dt>
        <dd>Quality filter an intercell network</dd>
      </dl><dl><dt>

          <code><a href="find_all_paths.html">find_all_paths()</a></code>

        </dt>
        <dd>All paths between two groups of vertices</dd>
      </dl><dl><dt>

          <code><a href="from_evidences.html">from_evidences()</a></code>

        </dt>
        <dd>Recreate interaction records from evidences columns</dd>
      </dl><dl><dt>

          <code><a href="get_db.html">get_db()</a></code>

        </dt>
        <dd>Access a built in database</dd>
      </dl><dl><dt>

          <code><a href="get_ontology_db.html">get_ontology_db()</a></code>

        </dt>
        <dd>Access an ontology database</dd>
      </dl><dl><dt>

          <code><a href="giant_component.html">giant_component()</a></code>

        </dt>
        <dd>Giant component of a graph</dd>
      </dl><dl><dt>

          <code><a href="go_annot_download.html">go_annot_download()</a></code>

        </dt>
        <dd>Gene annotations from Gene Ontology</dd>
      </dl><dl><dt>

          <code><a href="go_annot_slim.html">go_annot_slim()</a></code>

        </dt>
        <dd>GO slim gene annotations</dd>
      </dl><dl><dt>

          <code><a href="go_ontology_download.html">go_ontology_download()</a></code>

        </dt>
        <dd>The Gene Ontology tree</dd>
      </dl><dl><dt>

          <code><a href="graph_interaction.html">graph_interaction()</a></code>

        </dt>
        <dd>Interaction data frame from igraph graph object</dd>
      </dl><dl><dt>

          <code><a href="guide2pharma_download.html">guide2pharma_download()</a></code>

        </dt>
        <dd>Downloads interactions from the Guide to Pharmacology database</dd>
      </dl><dl><dt>

          <code><a href="harmonizome_download.html">harmonizome_download()</a></code>

        </dt>
        <dd>Downloads a Harmonizome network dataset</dd>
      </dl><dl><dt>

          <code><a href="has_extra_attrs.html">has_extra_attrs()</a></code>

        </dt>
        <dd>Tells if an interaction data frame has an extra_attrs column</dd>
      </dl><dl><dt>

          <code><a href="hmdb_id_mapping_table.html">hmdb_id_mapping_table()</a></code>

        </dt>
        <dd>Identifier translation table from HMDB</dd>
      </dl><dl><dt>

          <code><a href="hmdb_id_type.html">hmdb_id_type()</a></code>

        </dt>
        <dd>HMDB identifier type label</dd>
      </dl><dl><dt>

          <code><a href="hmdb_metabolite_fields.html">hmdb_metabolite_fields()</a></code>

        </dt>
        <dd>Field names for the HMDB metabolite dataset</dd>
      </dl><dl><dt>

          <code><a href="hmdb_protein_fields.html">hmdb_protein_fields()</a></code>

        </dt>
        <dd>Field names for the HMDB proteins dataset</dd>
      </dl><dl><dt>

          <code><a href="hmdb_table.html">hmdb_table()</a></code>

        </dt>
        <dd>Download a HMDB XML file and process it into a table</dd>
      </dl><dl><dt>

          <code><a href="homologene_download.html">homologene_download()</a></code>

        </dt>
        <dd>Orthology table for a pair of organisms</dd>
      </dl><dl><dt>

          <code><a href="homologene_organisms.html">homologene_organisms()</a></code>

        </dt>
        <dd>Organisms in NCBI HomoloGene</dd>
      </dl><dl><dt>

          <code><a href="homologene_raw.html">homologene_raw()</a></code>

        </dt>
        <dd>Orthology data from NCBI HomoloGene</dd>
      </dl><dl><dt>

          <code><a href="homologene_uniprot_orthology.html">homologene_uniprot_orthology()</a></code>

        </dt>
        <dd>Orthology table with UniProt IDs</dd>
      </dl><dl><dt>

          <code><a href="hpo_download.html">hpo_download()</a></code>

        </dt>
        <dd>Downloads protein annotations from Human Phenotype Ontology</dd>
      </dl><dl><dt>

          <code><a href="htridb_download.html">htridb_download()</a></code>

        </dt>
        <dd>Downloads TF-target interactions from HTRIdb</dd>
      </dl><dl><dt>

          <code><a href="id_translation_resources.html">id_translation_resources()</a></code>

        </dt>
        <dd>List available ID translation resources</dd>
      </dl><dl><dt>

          <code><a href="id_types.html">id_types()</a></code>

        </dt>
        <dd>ID types and synonyms in identifier translation</dd>
      </dl><dl><dt>

          <code><a href="inbiomap_download.html">inbiomap_download()</a></code>

        </dt>
        <dd>Downloads and preprocesses network data from InWeb InBioMap</dd>
      </dl><dl><dt>

          <code><a href="inbiomap_raw.html">inbiomap_raw()</a></code>

        </dt>
        <dd>Downloads network data from InWeb InBioMap</dd>
      </dl><dl><dt>

          <code><a href="interaction_datasets.html">interaction_datasets()</a></code>

        </dt>
        <dd>Datasets in the OmniPath Interactions database</dd>
      </dl><dl><dt>

          <code><a href="interaction_graph.html">interaction_graph()</a></code>

        </dt>
        <dd>Build Omnipath interaction graph</dd>
      </dl><dl><dt>

          <code><a href="interaction_resources.html">interaction_resources()</a></code>

        </dt>
        <dd>Interaction resources available in Omnipath</dd>
      </dl><dl><dt>

          <code><a href="interaction_types.html">interaction_types()</a></code>

        </dt>
        <dd>Interaction types in the OmniPath Interactions database</dd>
      </dl><dl><dt>

          <code><a href="intercell.html">intercell()</a></code>

        </dt>
        <dd>Cell-cell communication roles from OmniPath</dd>
      </dl><dl><dt>

          <code><a href="intercell_categories.html">intercell_categories()</a></code>

        </dt>
        <dd>Categories in the intercell database of OmniPath</dd>
      </dl><dl><dt>

          <code><a href="intercell_consensus_filter.html">intercell_consensus_filter()</a></code>

        </dt>
        <dd>Quality filter for intercell annotations</dd>
      </dl><dl><dt>

          <code><a href="intercell_generic_categories.html">intercell_generic_categories()</a></code>

        </dt>
        <dd>Retrieves a list of the generic categories in the intercell database of OmniPath</dd>
      </dl><dl><dt>

          <code><a href="intercell_network.html">intercell_network()</a></code>

        </dt>
        <dd>Intercellular communication network</dd>
      </dl><dl><dt>

          <code><a href="intercell_resources.html">intercell_resources()</a></code>

        </dt>
        <dd>Retrieves a list of intercellular communication resources available in OmniPath</dd>
      </dl><dl><dt>

          <code><a href="intercell_summary.html">intercell_summary()</a></code>

        </dt>
        <dd>Full list of intercell categories and resources</dd>
      </dl><dl><dt>

          <code><a href="is_ontology_id.html">is_ontology_id()</a></code>

        </dt>
        <dd>Looks like an ontology ID</dd>
      </dl><dl><dt>

          <code><a href="is_swissprot.html">is_swissprot()</a></code>

        </dt>
        <dd>Check for SwissProt IDs</dd>
      </dl><dl><dt>

          <code><a href="is_trembl.html">is_trembl()</a></code>

        </dt>
        <dd>Check for TrEMBL IDs</dd>
      </dl><dl><dt>

          <code><a href="is_uniprot.html">is_uniprot()</a></code>

        </dt>
        <dd>Looks like a UniProt ID?</dd>
      </dl><dl><dt>

          <code><a href="kegg_api_templates.html">kegg_api_templates()</a></code>

        </dt>
        <dd>List of templates in the KEGG REST API</dd>
      </dl><dl><dt>

          <code><a href="kegg_conv.html">kegg_conv()</a></code>

        </dt>
        <dd>Convert KEGG identifiers to/from outside identifiers</dd>
      </dl><dl><dt>

          <code><a href="kegg_databases.html">kegg_databases()</a></code>

        </dt>
        <dd>List of databases (endpoints) in the KEGG REST API</dd>
      </dl><dl><dt>

          <code><a href="kegg_ddi.html">kegg_ddi()</a></code>

        </dt>
        <dd>Find adverse drug-drug interactions in KEGG</dd>
      </dl><dl><dt>

          <code><a href="kegg_find.html">kegg_find()</a></code>

        </dt>
        <dd>Find entries in KEGG with matching query keyword or other query data</dd>
      </dl><dl><dt>

          <code><a href="kegg_info.html">kegg_info()</a></code>

        </dt>
        <dd>Information about a KEGG Pathway</dd>
      </dl><dl><dt>

          <code><a href="kegg_link.html">kegg_link()</a></code>

        </dt>
        <dd>Find related KEGG entries by using database cross-references</dd>
      </dl><dl><dt>

          <code><a href="kegg_list.html">kegg_list()</a></code>

        </dt>
        <dd>Obtain a list of KEGG entry identifiers and associated names</dd>
      </dl><dl><dt>

          <code><a href="kegg_open.html">kegg_open()</a></code>

        </dt>
        <dd>Open a KEGG Pathway diagram in the browser</dd>
      </dl><dl><dt>

          <code><a href="kegg_operations.html">kegg_operations()</a></code>

        </dt>
        <dd>List of operations in the KEGG REST API</dd>
      </dl><dl><dt>

          <code><a href="kegg_organism_codes.html">kegg_organism_codes()</a></code>

        </dt>
        <dd>All 3 letter organism code from KEGG</dd>
      </dl><dl><dt>

          <code><a href="kegg_organisms.html">kegg_organisms()</a></code>

        </dt>
        <dd>List of organisms in KEGG</dd>
      </dl><dl><dt>

          <code><a href="kegg_pathway_annotations.html">kegg_pathway_annotations()</a></code>

        </dt>
        <dd>Protein pathway annotations</dd>
      </dl><dl><dt>

          <code><a href="kegg_pathway_download.html">kegg_pathway_download()</a></code>

        </dt>
        <dd>Download one KEGG pathway</dd>
      </dl><dl><dt>

          <code><a href="kegg_pathway_list.html">kegg_pathway_list()</a></code>

        </dt>
        <dd>List of KEGG pathways</dd>
      </dl><dl><dt>

          <code><a href="kegg_pathways_download.html">kegg_pathways_download()</a></code>

        </dt>
        <dd>Download the KEGG Pathways database</dd>
      </dl><dl><dt>

          <code><a href="kegg_picture.html">kegg_picture()</a></code>

        </dt>
        <dd>Download a pathway diagram as a picture</dd>
      </dl><dl><dt>

          <code><a href="kegg_process.html">kegg_process()</a></code>

        </dt>
        <dd>Interactions from KGML</dd>
      </dl><dl><dt>

          <code><a href="kegg_query.html">kegg_query()</a></code>

        </dt>
        <dd>Compile a query for the KEGG REST API</dd>
      </dl><dl><dt>

          <code><a href="kegg_request.html">kegg_request()</a></code>

        </dt>
        <dd>Perform a KEGG REST API request</dd>
      </dl><dl><dt>

          <code><a href="kegg_rm_prefix.html">kegg_rm_prefix()</a></code>

        </dt>
        <dd>Remove prefix from KEGG foreign database identifiers</dd>
      </dl><dl><dt>

          <code><a href="latin_name.html">latin_name()</a></code>

        </dt>
        <dd>Latin (scientific) names of organisms</dd>
      </dl><dl><dt>

          <code><a href="load_db.html">load_db()</a></code>

        </dt>
        <dd>Load a built in database</dd>
      </dl><dl><dt>

          <code><a href="metalinksdb_sqlite.html">metalinksdb_sqlite()</a></code>

        </dt>
        <dd>Open MetalinksDB as an SQLite3 connection</dd>
      </dl><dl><dt>

          <code><a href="ncbi_taxid.html">ncbi_taxid()</a></code>

        </dt>
        <dd>NCBI Taxonomy IDs of organisms</dd>
      </dl><dl><dt>

          <code><a href="nichenet_build_model.html">nichenet_build_model()</a></code>

        </dt>
        <dd>Construct a NicheNet ligand-target model</dd>
      </dl><dl><dt>

          <code><a href="nichenet_expression_data.html">nichenet_expression_data()</a></code>

        </dt>
        <dd>Expression data from ligand-receptor perturbation experiments used by NicheNet</dd>
      </dl><dl><dt>

          <code><a href="nichenet_gr_network.html">nichenet_gr_network()</a></code>

        </dt>
        <dd>Builds a NicheNet gene regulatory network</dd>
      </dl><dl><dt>

          <code><a href="nichenet_gr_network_evex.html">nichenet_gr_network_evex()</a></code>

        </dt>
        <dd>NicheNet gene regulatory network from EVEX</dd>
      </dl><dl><dt>

          <code><a href="nichenet_gr_network_harmonizome.html">nichenet_gr_network_harmonizome()</a></code>

        </dt>
        <dd>NicheNet gene regulatory network from Harmonizome</dd>
      </dl><dl><dt>

          <code><a href="nichenet_gr_network_htridb.html">nichenet_gr_network_htridb()</a></code>

        </dt>
        <dd>NicheNet gene regulatory network from HTRIdb</dd>
      </dl><dl><dt>

          <code><a href="nichenet_gr_network_omnipath.html">nichenet_gr_network_omnipath()</a></code>

        </dt>
        <dd>Builds gene regulatory network for NicheNet using OmniPath</dd>
      </dl><dl><dt>

          <code><a href="nichenet_gr_network_pathwaycommons.html">nichenet_gr_network_pathwaycommons()</a></code>

        </dt>
        <dd>NicheNet gene regulatory network from PathwayCommons</dd>
      </dl><dl><dt>

          <code><a href="nichenet_gr_network_regnetwork.html">nichenet_gr_network_regnetwork()</a></code>

        </dt>
        <dd>NicheNet gene regulatory network from RegNetwork</dd>
      </dl><dl><dt>

          <code><a href="nichenet_gr_network_remap.html">nichenet_gr_network_remap()</a></code>

        </dt>
        <dd>NicheNet gene regulatory network from ReMap</dd>
      </dl><dl><dt>

          <code><a href="nichenet_gr_network_trrust.html">nichenet_gr_network_trrust()</a></code>

        </dt>
        <dd>NicheNet gene regulatory network from TRRUST</dd>
      </dl><dl><dt>

          <code><a href="nichenet_ligand_activities.html">nichenet_ligand_activities()</a></code>

        </dt>
        <dd>Calls the NicheNet ligand activity analysis</dd>
      </dl><dl><dt>

          <code><a href="nichenet_ligand_target_links.html">nichenet_ligand_target_links()</a></code>

        </dt>
        <dd>Compiles a table with weighted ligand-target links</dd>
      </dl><dl><dt>

          <code><a href="nichenet_ligand_target_matrix.html">nichenet_ligand_target_matrix()</a></code>

        </dt>
        <dd>Creates a NicheNet ligand-target matrix</dd>
      </dl><dl><dt>

          <code><a href="nichenet_lr_network.html">nichenet_lr_network()</a></code>

        </dt>
        <dd>Builds a NicheNet ligand-receptor network</dd>
      </dl><dl><dt>

          <code><a href="nichenet_lr_network_guide2pharma.html">nichenet_lr_network_guide2pharma()</a></code>

        </dt>
        <dd>Ligand-receptor network from Guide to Pharmacology</dd>
      </dl><dl><dt>

          <code><a href="nichenet_lr_network_omnipath.html">nichenet_lr_network_omnipath()</a></code>

        </dt>
        <dd>Builds ligand-receptor network for NicheNet using OmniPath</dd>
      </dl><dl><dt>

          <code><a href="nichenet_lr_network_ramilowski.html">nichenet_lr_network_ramilowski()</a></code>

        </dt>
        <dd>Ligand-receptor network from Ramilowski 2015</dd>
      </dl><dl><dt>

          <code><a href="nichenet_main.html">nichenet_main()</a></code>

        </dt>
        <dd>Executes the full NicheNet pipeline</dd>
      </dl><dl><dt>

          <code><a href="nichenet_networks.html">nichenet_networks()</a></code>

        </dt>
        <dd>Builds NicheNet network prior knowledge</dd>
      </dl><dl><dt>

          <code><a href="nichenet_optimization.html">nichenet_optimization()</a></code>

        </dt>
        <dd>Optimizes NicheNet model parameters</dd>
      </dl><dl><dt>

          <code><a href="nichenet_remove_orphan_ligands.html">nichenet_remove_orphan_ligands()</a></code>

        </dt>
        <dd>Removes experiments with orphan ligands</dd>
      </dl><dl><dt>

          <code><a href="nichenet_results_dir.html">nichenet_results_dir()</a></code>

        </dt>
        <dd>Path to the current NicheNet results directory</dd>
      </dl><dl><dt>

          <code><a href="nichenet_signaling_network.html">nichenet_signaling_network()</a></code>

        </dt>
        <dd>Builds a NicheNet signaling network</dd>
      </dl><dl><dt>

          <code><a href="nichenet_signaling_network_cpdb.html">nichenet_signaling_network_cpdb()</a></code>

        </dt>
        <dd>Builds signaling network for NicheNet using ConsensusPathDB</dd>
      </dl><dl><dt>

          <code><a href="nichenet_signaling_network_evex.html">nichenet_signaling_network_evex()</a></code>

        </dt>
        <dd>NicheNet signaling network from EVEX</dd>
      </dl><dl><dt>

          <code><a href="nichenet_signaling_network_harmonizome.html">nichenet_signaling_network_harmonizome()</a></code>

        </dt>
        <dd>NicheNet signaling network from Harmonizome</dd>
      </dl><dl><dt>

          <code><a href="nichenet_signaling_network_inbiomap.html">nichenet_signaling_network_inbiomap()</a></code>

        </dt>
        <dd>NicheNet signaling network from InWeb InBioMap</dd>
      </dl><dl><dt>

          <code><a href="nichenet_signaling_network_omnipath.html">nichenet_signaling_network_omnipath()</a></code>

        </dt>
        <dd>Builds signaling network for NicheNet using OmniPath</dd>
      </dl><dl><dt>

          <code><a href="nichenet_signaling_network_pathwaycommons.html">nichenet_signaling_network_pathwaycommons()</a></code>

        </dt>
        <dd>NicheNet signaling network from PathwayCommons</dd>
      </dl><dl><dt>

          <code><a href="nichenet_signaling_network_vinayagam.html">nichenet_signaling_network_vinayagam()</a></code>

        </dt>
        <dd>NicheNet signaling network from Vinayagam</dd>
      </dl><dl><dt>

          <code><a href="nichenet_test.html">nichenet_test()</a></code>

        </dt>
        <dd>Run the NicheNet pipeline with a little dummy network</dd>
      </dl><dl><dt>

          <code><a href="nichenet_workarounds.html">nichenet_workarounds()</a></code>

        </dt>
        <dd>Workarounds using NicheNet without attaching the package</dd>
      </dl><dl><dt>

          <code><a href="obo_parser.html">obo_parser()</a></code>

        </dt>
        <dd>Generic OBO parser</dd>
      </dl><dl><dt>

          <code><a href="oma_code.html">oma_code()</a></code>

        </dt>
        <dd>Orthologous Matrix (OMA) codes of organisms</dd>
      </dl><dl><dt>

          <code><a href="oma_organisms.html">oma_organisms()</a></code>

        </dt>
        <dd>Organism identifiers from the Orthologous Matrix</dd>
      </dl><dl><dt>

          <code><a href="oma_pairwise.html">oma_pairwise()</a></code>

        </dt>
        <dd>Orthologous gene pairs between two organisms</dd>
      </dl><dl><dt>

          <code><a href="oma_pairwise_genesymbols.html">oma_pairwise_genesymbols()</a></code>

        </dt>
        <dd>Orthologous pairs of gene symbols between two organisms</dd>
      </dl><dl><dt>

          <code><a href="oma_pairwise_translated.html">oma_pairwise_translated()</a></code>

        </dt>
        <dd>Orthologous pairs between two organisms for ID types not supported by OMA</dd>
      </dl><dl><dt>

          <code><a href="omnipath-interactions.html">omnipath_interactions()</a></code> <code><a href="omnipath-interactions.html">omnipath()</a></code> <code><a href="omnipath-interactions.html">pathwayextra()</a></code> <code><a href="omnipath-interactions.html">kinaseextra()</a></code> <code><a href="omnipath-interactions.html">ligrecextra()</a></code> <code><a href="omnipath-interactions.html">post_translational()</a></code> <code><a href="omnipath-interactions.html">dorothea()</a></code> <code><a href="omnipath-interactions.html">tf_target()</a></code> <code><a href="omnipath-interactions.html">transcriptional()</a></code> <code><a href="omnipath-interactions.html">collectri()</a></code> <code><a href="omnipath-interactions.html">mirna_target()</a></code> <code><a href="omnipath-interactions.html">tf_mirna()</a></code> <code><a href="omnipath-interactions.html">lncrna_mrna()</a></code> <code><a href="omnipath-interactions.html">small_molecule()</a></code> <code><a href="omnipath-interactions.html">all_interactions()</a></code>

        </dt>
        <dd>Molecular interactions from OmniPath</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_autoclean.html">omnipath_cache_autoclean()</a></code>

        </dt>
        <dd>Keeps only the latest versions of complete downloads</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_clean.html">omnipath_cache_clean()</a></code>

        </dt>
        <dd>Removes the items from the cache directory which are unknown by the cache database</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_clean_db.html">omnipath_cache_clean_db()</a></code>

        </dt>
        <dd>Removes the cache database entries without existing files</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_download_ready.html">omnipath_cache_download_ready()</a></code>

        </dt>
        <dd>Sets the download status to ready for a cache item</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_filter_versions.html">omnipath_cache_filter_versions()</a></code>

        </dt>
        <dd>Filters the versions from one cache record</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_get.html">omnipath_cache_get()</a></code>

        </dt>
        <dd>Retrieves one item from the cache directory</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_key.html">omnipath_cache_key()</a></code>

        </dt>
        <dd>Generates a hash which identifies an element in the cache database</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_latest_or_new.html">omnipath_cache_latest_or_new()</a></code>

        </dt>
        <dd>The latest or a new version of a cache record</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_latest_version.html">omnipath_cache_latest_version()</a></code>

        </dt>
        <dd>Finds the most recent version in a cache record</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_load.html">omnipath_cache_load()</a></code>

        </dt>
        <dd>Loads an R object from the cache</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_move_in.html">omnipath_cache_move_in()</a></code>

        </dt>
        <dd>Moves an existing file into the cache</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_remove.html">omnipath_cache_remove()</a></code>

        </dt>
        <dd>Removes contents from the cache directory</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_save.html">omnipath_cache_save()</a></code>

        </dt>
        <dd>Saves an R object to the cache</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_search.html">omnipath_cache_search()</a></code>

        </dt>
        <dd>Searches for cache items</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_set_ext.html">omnipath_cache_set_ext()</a></code>

        </dt>
        <dd>Sets the file extension for a cache record</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_update_status.html">omnipath_cache_update_status()</a></code>

        </dt>
        <dd>Updates the status of an existing cache record</dd>
      </dl><dl><dt>

          <code><a href="omnipath_cache_wipe.html">omnipath_cache_wipe()</a></code>

        </dt>
        <dd>Permanently removes all the cache contents</dd>
      </dl><dl><dt>

          <code><a href="omnipath_config_path.html">omnipath_config_path()</a></code> <code><a href="omnipath_config_path.html">config_path()</a></code>

        </dt>
        <dd>Current config file path of OmnipathR</dd>
      </dl><dl><dt>

          <code><a href="omnipath_for_cosmos.html">omnipath_for_cosmos()</a></code>

        </dt>
        <dd>OmniPath PPI for the COSMOS PKN</dd>
      </dl><dl><dt>

          <code><a href="omnipath_load_config.html">omnipath_load_config()</a></code> <code><a href="omnipath_load_config.html">load_config()</a></code>

        </dt>
        <dd>Load the package configuration from a config file</dd>
      </dl><dl><dt>

          <code><a href="omnipath_log.html">omnipath_log()</a></code> <code><a href="omnipath_log.html">read_log()</a></code>

        </dt>
        <dd>Browse the current OmnipathR log file</dd>
      </dl><dl><dt>

          <code><a href="omnipath_logfile.html">omnipath_logfile()</a></code> <code><a href="omnipath_logfile.html">logfile()</a></code>

        </dt>
        <dd>Path to the current OmnipathR log file</dd>
      </dl><dl><dt>

          <code><a href="omnipath_msg.html">omnipath_msg()</a></code>

        </dt>
        <dd>Dispatch a message to the OmnipathR logger</dd>
      </dl><dl><dt>

          <code><a href="omnipath_query.html">omnipath_query()</a></code>

        </dt>
        <dd>Download data from the OmniPath web service</dd>
      </dl><dl><dt>

          <code><a href="omnipath_reset_config.html">reset_config()</a></code> <code><a href="omnipath_reset_config.html">omnipath_reset_config()</a></code>

        </dt>
        <dd>Restore the built-in default values of all config parameters of a package</dd>
      </dl><dl><dt>

          <code><a href="omnipath_save_config.html">omnipath_save_config()</a></code> <code><a href="omnipath_save_config.html">save_config()</a></code>

        </dt>
        <dd>Save the current package configuration</dd>
      </dl><dl><dt>

          <code><a href="omnipath_set_cachedir.html">omnipath_set_cachedir()</a></code>

        </dt>
        <dd>Change the cache directory</dd>
      </dl><dl><dt>

          <code><a href="omnipath_set_console_loglevel.html">omnipath_set_console_loglevel()</a></code>

        </dt>
        <dd>Sets the log level for the console</dd>
      </dl><dl><dt>

          <code><a href="omnipath_set_logfile_loglevel.html">omnipath_set_logfile_loglevel()</a></code>

        </dt>
        <dd>Sets the log level for the logfile</dd>
      </dl><dl><dt>

          <code><a href="omnipath_set_loglevel.html">omnipath_set_loglevel()</a></code> <code><a href="omnipath_set_loglevel.html">set_loglevel()</a></code>

        </dt>
        <dd>Sets the log level for the package logger</dd>
      </dl><dl><dt>

          <code><a href="omnipath_show_db.html">omnipath_show_db()</a></code>

        </dt>
        <dd>Built in database definitions</dd>
      </dl><dl><dt>

          <code><a href="omnipath_unlock_cache_db.html">omnipath_unlock_cache_db()</a></code>

        </dt>
        <dd>Removes the lock file from the cache directory</dd>
      </dl><dl><dt>

          <code><a href="only_from.html">only_from()</a></code>

        </dt>
        <dd>Recreate interaction data frame based on certain datasets and resources</dd>
      </dl><dl><dt>

          <code><a href="ontology_ensure_id.html">ontology_ensure_id()</a></code>

        </dt>
        <dd>Only ontology IDs</dd>
      </dl><dl><dt>

          <code><a href="ontology_ensure_name.html">ontology_ensure_name()</a></code>

        </dt>
        <dd>Only ontology term names</dd>
      </dl><dl><dt>

          <code><a href="ontology_name_id.html">ontology_name_id()</a></code>

        </dt>
        <dd>Translate between ontology IDs and names</dd>
      </dl><dl><dt>

          <code><a href="organism_for.html">organism_for()</a></code>

        </dt>
        <dd>Make sure the resource supports the organism and it has the ID</dd>
      </dl><dl><dt>

          <code><a href="orthology_translate_column.html">orthology_translate_column()</a></code>

        </dt>
        <dd>Translate a column of identifiers by orthologous gene pairs</dd>
      </dl><dl><dt>

          <code><a href="pathwaycommons_download.html">pathwaycommons_download()</a></code>

        </dt>
        <dd>Interactions from PathwayCommons</dd>
      </dl><dl><dt>

          <code><a href="pivot_annotations.html">pivot_annotations()</a></code>

        </dt>
        <dd>Converts annotation tables to a wide format</dd>
      </dl><dl><dt>

          <code><a href="preppi_download.html">preppi_download()</a></code>

        </dt>
        <dd>Interactions from PrePPI</dd>
      </dl><dl><dt>

          <code><a href="preppi_filter.html">preppi_filter()</a></code>

        </dt>
        <dd>Filter PrePPI interactions by scores</dd>
      </dl><dl><dt>

          <code><a href="print_bma_motif_es.html">print_bma_motif_es()</a></code>

        </dt>
        <dd>Prints BMA motifs to the screen from a sequence of edges</dd>
      </dl><dl><dt>

          <code><a href="print_bma_motif_vs.html">print_bma_motif_vs()</a></code>

        </dt>
        <dd>Prints BMA motifs to the screen from a sequence of nodes</dd>
      </dl><dl><dt>

          <code><a href="print_interactions.html">print_interactions()</a></code>

        </dt>
        <dd>Print OmniPath interactions</dd>
      </dl><dl><dt>

          <code><a href="print_path_es.html">print_path_es()</a></code>

        </dt>
        <dd>Prints network paths in an edge sequence</dd>
      </dl><dl><dt>

          <code><a href="print_path_vs.html">print_path_vs()</a></code>

        </dt>
        <dd>Print networks paths given by node sequence</dd>
      </dl><dl><dt>

          <code><a href="pubmed_open.html">pubmed_open()</a></code>

        </dt>
        <dd>Open one or more PubMed articles</dd>
      </dl><dl><dt>

          <code><a href="query_info.html">query_info()</a></code>

        </dt>
        <dd>OmniPath query parameters</dd>
      </dl><dl><dt>

          <code><a href="ramilowski_download.html">ramilowski_download()</a></code>

        </dt>
        <dd>Downloads ligand-receptor interactions from Ramilowski et al. 2015</dd>
      </dl><dl><dt>

          <code><a href="ramp_id_mapping_table.html">ramp_id_mapping_table()</a></code>

        </dt>
        <dd>Pairwise ID translation table from RaMP database</dd>
      </dl><dl><dt>

          <code><a href="ramp_id_type.html">ramp_id_type()</a></code>

        </dt>
        <dd>RaMP identifier type label</dd>
      </dl><dl><dt>

          <code><a href="ramp_sqlite.html">ramp_sqlite()</a></code>

        </dt>
        <dd>Download and open RaMP database SQLite</dd>
      </dl><dl><dt>

          <code><a href="ramp_table.html">ramp_table()</a></code>

        </dt>
        <dd>Return table from RaMP database</dd>
      </dl><dl><dt>

          <code><a href="ramp_tables.html">ramp_tables()</a></code>

        </dt>
        <dd>List tables in RaMP database</dd>
      </dl><dl><dt>

          <code><a href="regnetwork_directions.html">regnetwork_directions()</a></code>

        </dt>
        <dd>Transcription factor effects from RegNetwork</dd>
      </dl><dl><dt>

          <code><a href="regnetwork_download.html">regnetwork_download()</a></code>

        </dt>
        <dd>Interactions from RegNetwork</dd>
      </dl><dl><dt>

          <code><a href="relations_list_to_table.html">relations_list_to_table()</a></code>

        </dt>
        <dd>Table from a nested list of ontology relations</dd>
      </dl><dl><dt>

          <code><a href="relations_table_to_graph.html">relations_table_to_graph()</a></code>

        </dt>
        <dd>Graph from a table of ontology relations</dd>
      </dl><dl><dt>

          <code><a href="relations_table_to_list.html">relations_table_to_list()</a></code>

        </dt>
        <dd>Nested list from a table of ontology relations</dd>
      </dl><dl><dt>

          <code><a href="remap_dorothea_download.html">remap_dorothea_download()</a></code>

        </dt>
        <dd>Downloads TF-target interactions from ReMap</dd>
      </dl><dl><dt>

          <code><a href="remap_filtered.html">remap_filtered()</a></code>

        </dt>
        <dd>Downloads TF-target interactions from ReMap</dd>
      </dl><dl><dt>

          <code><a href="remap_tf_target_download.html">remap_tf_target_download()</a></code>

        </dt>
        <dd>Downloads TF-target interactions from ReMap</dd>
      </dl><dl><dt>

          <code><a href="resource_info.html">resource_info()</a></code>

        </dt>
        <dd>OmniPath resource information</dd>
      </dl><dl><dt>

          <code><a href="resources.html">resources()</a></code>

        </dt>
        <dd>Retrieve the available resources for a given query type</dd>
      </dl><dl><dt>

          <code><a href="resources_colname.html">resources_colname()</a></code>

        </dt>
        <dd>Name of the column with the resources</dd>
      </dl><dl><dt>

          <code><a href="resources_in.html">resources_in()</a></code>

        </dt>
        <dd>Collect resource names from a data frame</dd>
      </dl><dl><dt>

          <code><a href="show_network.html">show_network()</a></code>

        </dt>
        <dd>Visualize node neighborhood with SigmaJS</dd>
      </dl><dl><dt>

          <code><a href="signed_ptms.html">signed_ptms()</a></code>

        </dt>
        <dd>Causal effect enzyme-PTM interactions</dd>
      </dl><dl><dt>

          <code><a href="simplify_intercell_network.html">simplify_intercell_network()</a></code>

        </dt>
        <dd>Simplify an intercell network</dd>
      </dl><dl><dt>

          <code><a href="static_table.html">static_table()</a></code>

        </dt>
        <dd>Retrieve a static table from OmniPath</dd>
      </dl><dl><dt>

          <code><a href="static_tables.html">static_tables()</a></code>

        </dt>
        <dd>List the static tables available from OmniPath</dd>
      </dl><dl><dt>

          <code><a href="stitch_actions.html">stitch_actions()</a></code>

        </dt>
        <dd>Retrieve the STITCH actions dataset</dd>
      </dl><dl><dt>

          <code><a href="stitch_links.html">stitch_links()</a></code>

        </dt>
        <dd>Retrieve the STITCH links dataset</dd>
      </dl><dl><dt>

          <code><a href="stitch_network.html">stitch_network()</a></code>

        </dt>
        <dd>Chemical-protein interactions from STITCH</dd>
      </dl><dl><dt>

          <code><a href="stitch_remove_prefixes.html">stitch_remove_prefixes()</a></code>

        </dt>
        <dd>Remove the prefixes from STITCH identifiers</dd>
      </dl><dl><dt>

          <code><a href="subnetwork.html">subnetwork()</a></code>

        </dt>
        <dd>Extract a custom subnetwork from a large network</dd>
      </dl><dl><dt>

          <code><a href="swap_relations.html">swap_relations()</a></code>

        </dt>
        <dd>Reverse the direction of ontology relations</dd>
      </dl><dl><dt>

          <code><a href="swissprots_only.html">swissprots_only()</a></code>

        </dt>
        <dd>Retain only SwissProt IDs</dd>
      </dl><dl><dt>

          <code><a href="tfcensus_download.html">tfcensus_download()</a></code>

        </dt>
        <dd>Downloads the list of transcription factors from TF census</dd>
      </dl><dl><dt>

          <code><a href="translate_ids.html">translate_ids()</a></code>

        </dt>
        <dd>Translate gene, protein and small molecule identifiers</dd>
      </dl><dl><dt>

          <code><a href="translate_ids_multi.html">translate_ids_multi()</a></code>

        </dt>
        <dd>Translate gene, protein and small molecule identifiers from multiple columns</dd>
      </dl><dl><dt>

          <code><a href="trembls_only.html">trembls_only()</a></code>

        </dt>
        <dd>Retain only TrEMBL IDs</dd>
      </dl><dl><dt>

          <code><a href="trrust_download.html">trrust_download()</a></code>

        </dt>
        <dd>Downloads TF-target interactions from TRRUST</dd>
      </dl><dl><dt>

          <code><a href="uniprot_full_id_mapping_table.html">uniprot_full_id_mapping_table()</a></code>

        </dt>
        <dd>Creates an ID translation table from UniProt data</dd>
      </dl><dl><dt>

          <code><a href="uniprot_genesymbol_cleanup.html">uniprot_genesymbol_cleanup()</a></code>

        </dt>
        <dd>TrEMBL to SwissProt by gene names</dd>
      </dl><dl><dt>

          <code><a href="uniprot_id_mapping_table.html">uniprot_id_mapping_table()</a></code>

        </dt>
        <dd>ID translation data from UniProt ID Mapping</dd>
      </dl><dl><dt>

          <code><a href="uniprot_id_type.html">uniprot_id_type()</a></code>

        </dt>
        <dd>UniProt identifier type label</dd>
      </dl><dl><dt>

          <code><a href="uniprot_idmapping_id_types.html">uniprot_idmapping_id_types()</a></code>

        </dt>
        <dd>ID types available in the UniProt ID Mapping service</dd>
      </dl><dl><dt>

          <code><a href="unique_intercell_network.html">unique_intercell_network()</a></code>

        </dt>
        <dd>Unique intercellular interactions</dd>
      </dl><dl><dt>

          <code><a href="unnest_evidences.html">unnest_evidences()</a></code>

        </dt>
        <dd>Separate evidences by direction and effect sign</dd>
      </dl><dl><dt>

          <code><a href="uploadlists_id_type.html">uploadlists_id_type()</a></code>

        </dt>
        <dd>UniProt Uploadlists identifier type label</dd>
      </dl><dl><dt>

          <code><a href="vinayagam_download.html">vinayagam_download()</a></code>

        </dt>
        <dd>Protein-protein interactions from Vinayagam 2011</dd>
      </dl><dl><dt>

          <code><a href="walk_ontology_tree.html">walk_ontology_tree()</a></code>

        </dt>
        <dd>All nodes of a subtree starting from the selected nodes</dd>
      </dl><dl><dt>

          <code><a href="with_extra_attrs.html">with_extra_attrs()</a></code>

        </dt>
        <dd>Interaction records having certain extra attributes</dd>
      </dl><dl><dt>

          <code><a href="with_references.html">with_references()</a></code>

        </dt>
        <dd>Interactions having references</dd>
      </dl><dl><dt>

          <code><a href="zenodo_download.html">zenodo_download()</a></code>

        </dt>
        <dd>Retrieves data from Zenodo</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Valdeolivas, Denes Turei, Attila Gabor, Diego Mananes, Aurelien Dugourd.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

