<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Orthologous pairs between two organisms for ID types not supported by OMA — oma_pairwise_translated • OmnipathR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Orthologous pairs between two organisms for ID types not supported by OMA — oma_pairwise_translated"><meta name="description" content="The Orthologous Matrix (OMA), a resource of orthologous relationships
between genes, doesn't provide gene symbols, the identifier preferred in
many bioinformatics pipelines. Hence this function wraps
oma_pairwise by translating the identifiers used in OMA to
gene symbols. Items that can not be translated to `id_type` (but present
in the data with their internal OMA IDs) will be removed. Then,
in this function we translate the identifiers to the desired ID type."><meta property="og:description" content="The Orthologous Matrix (OMA), a resource of orthologous relationships
between genes, doesn't provide gene symbols, the identifier preferred in
many bioinformatics pipelines. Hence this function wraps
oma_pairwise by translating the identifiers used in OMA to
gene symbols. Items that can not be translated to `id_type` (but present
in the data with their internal OMA IDs) will be removed. Then,
in this function we translate the identifiers to the desired ID type."><meta property="og:image" content="https://saezlab.github.io/OmnipathR/logo.svg"><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-10"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-10');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">OmnipathR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.12.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bioc_workshop.html">OmniPath Bioconductor workshop</a></li>
    <li><a class="dropdown-item" href="../articles/db_manager.html">The database manager in OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/drug_targets.html">Building protein networks around drug-targets using OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/extra_attrs.html">Resource specific interaction attributes</a></li>
    <li><a class="dropdown-item" href="../articles/gene_ontology.html">Gene Ontology in OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/nichenet.html">Using NicheNet with OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/omnipath_intro.html">OmnipathR: an R client for the OmniPath web service</a></li>
    <li><a class="dropdown-item" href="../articles/paths.html">Pathway construction</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/saezlab/OmnipathR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/omnipathdb"><span class="fa fab fa-twitter fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://omnipathdb.org/"><span class="fa fas fa-university"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Orthologous pairs between two organisms for ID types not supported by OMA</h1>
      <small class="dont-index">Source: <a href="https://github.com/saezlab/OmnipathR/blob/HEAD/R/oma.R" class="external-link"><code>R/oma.R</code></a></small>
      <div class="d-none name"><code>oma_pairwise_translated.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The Orthologous Matrix (OMA), a resource of orthologous relationships
between genes, doesn't provide gene symbols, the identifier preferred in
many bioinformatics pipelines. Hence this function wraps
<code><a href="oma_pairwise.html">oma_pairwise</a></code> by translating the identifiers used in OMA to
gene symbols. Items that can not be translated to `id_type` (but present
in the data with their internal OMA IDs) will be removed. Then,
in this function we translate the identifiers to the desired ID type.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">oma_pairwise_translated</span><span class="op">(</span></span>
<span>  organism_a <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>  organism_b <span class="op">=</span> <span class="st">"mouse"</span>,</span>
<span>  id_type <span class="op">=</span> <span class="st">"uniprot"</span>,</span>
<span>  oma_id_type <span class="op">=</span> <span class="st">"uniprot_entry"</span>,</span>
<span>  mappings <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1:1"</span>, <span class="st">"1:m"</span>, <span class="st">"n:1"</span>, <span class="st">"n:m"</span><span class="op">)</span>,</span>
<span>  only_ids <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-organism-a">organism_a<a class="anchor" aria-label="anchor" href="#arg-organism-a"></a></dt>
<dd><p>Name or identifier of an organism.</p></dd>


<dt id="arg-organism-b">organism_b<a class="anchor" aria-label="anchor" href="#arg-organism-b"></a></dt>
<dd><p>Name or identifier of another organism.</p></dd>


<dt id="arg-id-type">id_type<a class="anchor" aria-label="anchor" href="#arg-id-type"></a></dt>
<dd><p>The gene or protein identifier to use in the table. For a
list of supported ID types see `omnipath.env$id_types$oma`. These are
the identifiers that will be translated to gene symbols.</p></dd>


<dt id="arg-oma-id-type">oma_id_type<a class="anchor" aria-label="anchor" href="#arg-oma-id-type"></a></dt>
<dd><p>Character: the gene or protein identifier to be queried
from OMA. These IDs will be translated to `id_type`.</p></dd>


<dt id="arg-mappings">mappings<a class="anchor" aria-label="anchor" href="#arg-mappings"></a></dt>
<dd><p>Character vector: control ambiguous mappings:</p><ul><li><p>1:1 - unambiguous</p></li>
<li><p>1:m - one-to-many</p></li>
<li><p>n:1 - many-to-one</p></li>
<li><p>n:m - many-to-many</p></li>
</ul></dd>


<dt id="arg-only-ids">only_ids<a class="anchor" aria-label="anchor" href="#arg-only-ids"></a></dt>
<dd><p>Logical: include only the two identifier columns, not the
mapping type and the orthology group columns.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with orthologous gene pairs.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">oma_pairwise_translated</span><span class="op">(</span><span class="st">"human"</span>, <span class="st">"mouse"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 13,006 × 2</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    id_organism_a id_organism_b</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> P21452        P30549       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> Q8TAD7        P0C913       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> Q8N392        Q8K0Q5       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> O60285        Q641K5       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> Q07065        Q8BMK4       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> P78371        P80314       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Q8N4U5        Q8K1H7       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> Q8WU20        Q8C180       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> O95619        Q9CR11       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> Q9NW08        P59470       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 12,996 more rows</span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Valdeolivas, Denes Turei, Attila Gabor.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

