<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Molecular interactions from OmniPath — omnipath-interactions • OmnipathR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Molecular interactions from OmniPath — omnipath-interactions"><meta name="description" content='The functions listed here all download pairwise, causal molecular
interactions from the https://omnipathdb.org/interactions endpoint of
the OmniPath web service. They are different only in the type of
interactions and the kind of resources and data they have been compiled
from. A complete list of these functions is available below, these cover
the interaction datasets and types  currently available in OmniPath:
Interactions from the https://omnipathdb.org/interactions endpoint of
the OmniPath web service.
By default, it downloads only the "omnipath" dataset, which corresponds to
the curated causal interactions described in Turei et al. 2016.
Imports interactions from the `omnipath` dataset of OmniPath, a dataset
that inherits most of its design and contents from the original OmniPath
core from the 2016 publication. This dataset consists of about 40k
interactions.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=pathwayextra,
which contains activity flow interactions without literature reference.
The activity flow interactions supported by literature references
are part of the `omnipath` dataset.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=kinaseextra,
which contains enzyme-substrate interactions without literature reference.
The enzyme-substrate interactions supported by literature references
are part of the `omnipath` dataset.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=ligrecextra,
which contains ligand-receptor interactions without literature reference.
The ligand-receptor interactions supported by literature references
are part of the `omnipath` dataset.
Imports interactions from all post-translational datasets of OmniPath.
The datasets are "omnipath", "kinaseextra", "pathwayextra" and
"ligrecextra".
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=dorothea
which contains transcription factor (TF)-target interactions from DoRothEA
https://github.com/saezlab/DoRothEA
DoRothEA is a comprehensive resource of transcriptional regulation,
consisting of 16 original resources, in silico TFBS prediction, gene
expression signatures and ChIP-Seq binding site analysis.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=tf_target,
which contains transcription factor-target protein coding gene
interactions. Note: this is not the only TF-target dataset in OmniPath,
`dorothea` is the other one and the `tf_mirna` dataset provides
TF-miRNA gene interactions.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=tf_target,dorothea,
which contains transcription factor-target protein coding gene
interactions.
CollecTRI is a comprehensive resource of transcriptional regulation,
published in 2023, consisting of 14 resources and original literature
curation.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=mirnatarget,
which contains miRNA-mRNA interactions.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=tf_mirna,
which contains transcription factor-miRNA gene interactions
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=lncrna_mrna,
which contains lncRNA-mRNA interactions
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=small_molecule,
which contains small molecule-protein interactions. Small molecules
can be metabolites, intrinsic ligands or drug compounds.'><meta property="og:description" content='The functions listed here all download pairwise, causal molecular
interactions from the https://omnipathdb.org/interactions endpoint of
the OmniPath web service. They are different only in the type of
interactions and the kind of resources and data they have been compiled
from. A complete list of these functions is available below, these cover
the interaction datasets and types  currently available in OmniPath:
Interactions from the https://omnipathdb.org/interactions endpoint of
the OmniPath web service.
By default, it downloads only the "omnipath" dataset, which corresponds to
the curated causal interactions described in Turei et al. 2016.
Imports interactions from the `omnipath` dataset of OmniPath, a dataset
that inherits most of its design and contents from the original OmniPath
core from the 2016 publication. This dataset consists of about 40k
interactions.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=pathwayextra,
which contains activity flow interactions without literature reference.
The activity flow interactions supported by literature references
are part of the `omnipath` dataset.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=kinaseextra,
which contains enzyme-substrate interactions without literature reference.
The enzyme-substrate interactions supported by literature references
are part of the `omnipath` dataset.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=ligrecextra,
which contains ligand-receptor interactions without literature reference.
The ligand-receptor interactions supported by literature references
are part of the `omnipath` dataset.
Imports interactions from all post-translational datasets of OmniPath.
The datasets are "omnipath", "kinaseextra", "pathwayextra" and
"ligrecextra".
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=dorothea
which contains transcription factor (TF)-target interactions from DoRothEA
https://github.com/saezlab/DoRothEA
DoRothEA is a comprehensive resource of transcriptional regulation,
consisting of 16 original resources, in silico TFBS prediction, gene
expression signatures and ChIP-Seq binding site analysis.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=tf_target,
which contains transcription factor-target protein coding gene
interactions. Note: this is not the only TF-target dataset in OmniPath,
`dorothea` is the other one and the `tf_mirna` dataset provides
TF-miRNA gene interactions.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=tf_target,dorothea,
which contains transcription factor-target protein coding gene
interactions.
CollecTRI is a comprehensive resource of transcriptional regulation,
published in 2023, consisting of 14 resources and original literature
curation.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=mirnatarget,
which contains miRNA-mRNA interactions.
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=tf_mirna,
which contains transcription factor-miRNA gene interactions
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=lncrna_mrna,
which contains lncRNA-mRNA interactions
Imports the dataset from:
https://omnipathdb.org/interactions?datasets=small_molecule,
which contains small molecule-protein interactions. Small molecules
can be metabolites, intrinsic ligands or drug compounds.'><meta property="og:image" content="https://saezlab.github.io/OmnipathR/logo.svg"><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-10"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-10');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">OmnipathR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.15.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bioc_workshop.html">OmniPath Bioconductor workshop</a></li>
    <li><a class="dropdown-item" href="../articles/cosmos.html">Building prior knowledge network (PKN) for COSMOS</a></li>
    <li><a class="dropdown-item" href="../articles/db_manager.html">The database manager in OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/drug_targets.html">Building protein networks around drug-targets using OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/extra_attrs.html">Resource specific interaction attributes</a></li>
    <li><a class="dropdown-item" href="../articles/gene_ontology.html">Gene Ontology in OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/nichenet.html">Using NicheNet with OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/omnipath_intro.html">OmnipathR: an R client for the OmniPath web service</a></li>
    <li><a class="dropdown-item" href="../articles/paths.html">Pathway construction</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/saezlab/OmnipathR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/omnipathdb" aria-label="Go to the Twitter of OmniPath"><span class="fa fab fa-twitter fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://omnipathdb.org/" aria-label="Go to the OmniPath homepage"><span class="fa fas fa-university"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Molecular interactions from OmniPath</h1>
      <small class="dont-index">Source: <a href="https://github.com/saezlab/OmnipathR/blob/HEAD/R/interactions.R" class="external-link"><code>R/interactions.R</code></a></small>
      <div class="d-none name"><code>omnipath-interactions.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The functions listed here all download pairwise, causal molecular
interactions from the <a href="https://omnipathdb.org/interactions" class="external-link">https://omnipathdb.org/interactions</a> endpoint of
the OmniPath web service. They are different only in the type of
interactions and the kind of resources and data they have been compiled
from. A complete list of these functions is available below, these cover
the interaction datasets and types  currently available in OmniPath:</p>
<p>Interactions from the <a href="https://omnipathdb.org/interactions" class="external-link">https://omnipathdb.org/interactions</a> endpoint of
the OmniPath web service.
By default, it downloads only the "omnipath" dataset, which corresponds to
the curated causal interactions described in Turei et al. 2016.</p>
<p>Imports interactions from the `omnipath` dataset of OmniPath, a dataset
that inherits most of its design and contents from the original OmniPath
core from the 2016 publication. This dataset consists of about 40k
interactions.</p>
<p>Imports the dataset from:
<a href="https://omnipathdb.org/interactions?datasets=pathwayextra" class="external-link">https://omnipathdb.org/interactions?datasets=pathwayextra</a>,
which contains activity flow interactions without literature reference.
The activity flow interactions supported by literature references
are part of the `omnipath` dataset.</p>
<p>Imports the dataset from:
<a href="https://omnipathdb.org/interactions?datasets=kinaseextra" class="external-link">https://omnipathdb.org/interactions?datasets=kinaseextra</a>,
which contains enzyme-substrate interactions without literature reference.
The enzyme-substrate interactions supported by literature references
are part of the `omnipath` dataset.</p>
<p>Imports the dataset from:
<a href="https://omnipathdb.org/interactions?datasets=ligrecextra" class="external-link">https://omnipathdb.org/interactions?datasets=ligrecextra</a>,
which contains ligand-receptor interactions without literature reference.
The ligand-receptor interactions supported by literature references
are part of the `omnipath` dataset.</p>
<p>Imports interactions from all post-translational datasets of OmniPath.
The datasets are "omnipath", "kinaseextra", "pathwayextra" and
"ligrecextra".</p>
<p>Imports the dataset from:
<a href="https://omnipathdb.org/interactions?datasets=dorothea" class="external-link">https://omnipathdb.org/interactions?datasets=dorothea</a>
which contains transcription factor (TF)-target interactions from DoRothEA
<a href="https://github.com/saezlab/DoRothEA" class="external-link">https://github.com/saezlab/DoRothEA</a>
DoRothEA is a comprehensive resource of transcriptional regulation,
consisting of 16 original resources, in silico TFBS prediction, gene
expression signatures and ChIP-Seq binding site analysis.</p>
<p>Imports the dataset from:
<a href="https://omnipathdb.org/interactions?datasets=tf_target" class="external-link">https://omnipathdb.org/interactions?datasets=tf_target</a>,
which contains transcription factor-target protein coding gene
interactions. Note: this is not the only TF-target dataset in OmniPath,
`dorothea` is the other one and the `tf_mirna` dataset provides
TF-miRNA gene interactions.</p>
<p>Imports the dataset from:
<a href="https://omnipathdb.org/interactions?datasets=tf_target,dorothea" class="external-link">https://omnipathdb.org/interactions?datasets=tf_target,dorothea</a>,
which contains transcription factor-target protein coding gene
interactions.</p>
<p>CollecTRI is a comprehensive resource of transcriptional regulation,
published in 2023, consisting of 14 resources and original literature
curation.</p>
<p>Imports the dataset from:
<a href="https://omnipathdb.org/interactions?datasets=mirnatarget" class="external-link">https://omnipathdb.org/interactions?datasets=mirnatarget</a>,
which contains miRNA-mRNA interactions.</p>
<p>Imports the dataset from:
<a href="https://omnipathdb.org/interactions?datasets=tf_mirna" class="external-link">https://omnipathdb.org/interactions?datasets=tf_mirna</a>,
which contains transcription factor-miRNA gene interactions</p>
<p>Imports the dataset from:
<a href="https://omnipathdb.org/interactions?datasets=lncrna_mrna" class="external-link">https://omnipathdb.org/interactions?datasets=lncrna_mrna</a>,
which contains lncRNA-mRNA interactions</p>
<p>Imports the dataset from:
<a href="https://omnipathdb.org/interactions?datasets=small_molecule" class="external-link">https://omnipathdb.org/interactions?datasets=small_molecule</a>,
which contains small molecule-protein interactions. Small molecules
can be metabolites, intrinsic ligands or drug compounds.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">omnipath_interactions</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">omnipath</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pathwayextra</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">kinaseextra</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ligrecextra</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">post_translational</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dorothea</span><span class="op">(</span>dorothea_levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">tf_target</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">transcriptional</span><span class="op">(</span>dorothea_levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">collectri</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">mirna_target</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">tf_mirna</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lncrna_mrna</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">small_molecule</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">all_interactions</span><span class="op">(</span></span>
<span>  dorothea_levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>  types <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fields <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Arguments passed on to <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code>, <code><a href="omnipath_query.html">omnipath_query</a></code></p><dl><dt><code>organism</code></dt>
<dd><p>Character or integer: name or NCBI Taxonomy ID of the
organism. OmniPath is built of human data, and the web service provides
orthology translated interactions and enzyme-substrate relationships for
mouse and rat. For other organisms and query types, orthology
translation will be called automatically on the downloaded human data
before returning the result.</p></dd>

    <dt><code>resources</code></dt>
<dd><p>Character vector: name of one or more resources. Restrict
the data to these resources. For a complete list of available resources,
call the `&lt;query_type&gt;_resources` functions for the query type of
interst.</p></dd>

    <dt><code>datasets</code></dt>
<dd><p>Character vector: name of one or more datasets. In the
interactions query type a number of datasets are available. The default
is caled "omnipath", and corresponds to the curated causal signaling
network published in the 2016 OmniPath paper.</p></dd>

    <dt><code>genesymbols</code></dt>
<dd><p>Character or logical: TRUE or FALS or "yes" or "no".
Include the `genesymbols` column in the results. OmniPath uses UniProt
IDs as the primary identifiers, gene symbols are optional.</p></dd>

    <dt><code>default_fields</code></dt>
<dd><p>Logical: if TRUE, the default fields will be included.</p></dd>

    <dt><code>silent</code></dt>
<dd><p>Logical: if TRUE, no messages will be printed. By default a
summary message is printed upon successful download.</p></dd>

    <dt><code>logicals</code></dt>
<dd><p>Character vector: fields to be cast to logical.</p></dd>

    <dt><code>format</code></dt>
<dd><p>Character: if "json", JSON will be retrieved and processed
into a nested list; any other value will return data frame.</p></dd>

    <dt><code>download_args</code></dt>
<dd><p>List: parameters to pass to the download function,
which is `readr::read_tsv` by default, and `jsonlite::safe_load`.</p></dd>

    <dt><code>references_by_resource</code></dt>
<dd><p>Logical: if TRUE,, in the `references`
column the PubMed IDs will be prefixed with the names of the resources
they are coming from. If FALSE, the `references` column will be a list
of unique PubMed IDs.</p></dd>

    <dt><code>add_counts</code></dt>
<dd><p>Logical: if TRUE, the number of references and number of
resources for each record will be added to the result.</p></dd>

    <dt><code>license</code></dt>
<dd><p>Character: license restrictions. By default, data from
resources allowing "academic" use is returned by OmniPath. If you use
the data for work in a company, you can provide "commercial" or
"for-profit", which will restrict the data to those records which are
supported by resources that allow for-profit use.</p></dd>

    <dt><code>password</code></dt>
<dd><p>Character: password for the OmniPath web service. You can
provide a special password here which enables the use of `license =
"ignore"` option, completely bypassing the license filter.</p></dd>

    <dt><code>json_param</code></dt>
<dd><p>List: parameters to pass to the `jsonlite::fromJSON` when
processing JSON columns embedded in the downloaded data. Such columns
are "extra_attrs" and "evidences". These are optional columns which
provide a lot of extra details about interactions.</p></dd>

    <dt><code>strict_evidences</code></dt>
<dd><p>Logical: reconstruct the "sources" and "references"
columns of interaction data frames based on the "evidences" column,
strictly filtering them to the queried datasets and resources. Without
this, the "sources" and "references" fields for each record might
contain information for datasets and resources other than the queried
ones, because the downloaded records are a result of a simple filtering
of an already integrated data frame.</p></dd>

    <dt><code>genesymbol_resource</code></dt>
<dd><p>Character: "uniprot" (default) or "ensembl". The
OmniPath web service uses the primary gene symbols as provided by
UniProt. By passing "ensembl" here, the UniProt gene symbols will be
replaced by the ones used in Ensembl. This translation results in a loss
of a few records, and multiplication of another few records due to
ambiguous translation.</p></dd>

    <dt><code>cache</code></dt>
<dd><p>Logical: use caching, load data from and save to the. The cache
directory by default belongs to the user, located in the user's default
cache directory, and named "OmnipathR". Find out about it by
<code>getOption("omnipathr.cachedir")</code>. Can be changed by
<code><a href="omnipath_set_cachedir.html">omnipath_set_cachedir</a></code>.</p></dd>


</dl></dd>


<dt id="arg-dorothea-levels">dorothea_levels<a class="anchor" aria-label="anchor" href="#arg-dorothea-levels"></a></dt>
<dd><p>The confidence levels of the dorothea
interactions (TF-target) which range from A to D. Set to A and B by
default.</p></dd>


<dt id="arg-types">types<a class="anchor" aria-label="anchor" href="#arg-types"></a></dt>
<dd><p>Character: interaction types, such as "transcriptional",
"post_transcriptional", "post_translational", etc.</p></dd>


<dt id="arg-fields">fields<a class="anchor" aria-label="anchor" href="#arg-fields"></a></dt>
<dd><p>Character: additional fields (columns) to be included in the
result. For a list of available fields, see <code><a href="query_info.html">query_info</a></code>.</p></dd>


<dt id="arg-exclude">exclude<a class="anchor" aria-label="anchor" href="#arg-exclude"></a></dt>
<dd><p>Character: names of datasets or resource to be excluded from
the result. By deafult, the records supported by only these resources or
datasets will be removed from the output. If <code>strict_evidences =
TRUE</code>, the resource, reference and causality information in the data
frame will be reconstructed to remove all information coming from the
excluded resources.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A dataframe of molecular interactions.</p>
<p>A dataframe of literature curated, post-translational signaling
    interactions.</p>
<p>A dataframe containing activity flow interactions between proteins
without literature reference</p>
<p>A dataframe containing enzyme-substrate interactions without
literature reference</p>
<p>A dataframe containing ligand-receptor interactions including
the ones without literature references</p>
<p>A dataframe containing post-translational interactions</p>
<p>A data frame of TF-target interactions from DoRothEA.</p>
<p>A dataframe containing TF-target interactions</p>
<p>A dataframe containing TF-target interactions.</p>
<p>A dataframe of TF-target interactions.</p>
<p>A dataframe containing miRNA-mRNA interactions</p>
<p>A dataframe containing TF-miRNA interactions</p>
<p>A dataframe containing lncRNA-mRNA interactions</p>
<p>A dataframe of small molecule-protein interactions</p>
<p>A dataframe containing all the datasets in the interactions query</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><strong>Post-translational (protein-protein, PPI) interactions</strong></p>
<ul><li><p><code>omnipath</code>: the OmniPath data as defined in the 2016 paper,
an arbitrary optimum between coverage and quality. This dataset
contains almost entirely causal (stimulatory or inhibitory; i.e.
activity flow , according to the SBGN standard), physical
interactions between pairs of proteins, curated by experts
from the literature.</p></li>
<li><p><code>pathwayextra</code>: activity flow interactions without literature
references.</p></li>
<li><p><code>kinaseextra</code>: enzyme-substrate interactions without
literature references.</p></li>
<li><p><code>ligrecextra</code>: ligand-receptor interactions without
literature references.</p></li>
<li><p><code>post_translational</code>: all post-translational
(protein-protein, PPI) interactions; this is the combination of the
<em>omnipath</em>, <em>pathwayextra</em>, <em>kinaseextra</em> and <em>ligrecextra</em>
datasets.</p></li>
</ul><p><strong>TF-target (gene regulatory, GRN) interactions</strong></p>
<ul><li><p><code>collectri</code>: transcription factor (TF)-target
interactions from CollecTRI.</p></li>
<li><p><code>dorothea</code>: transcription factor (TF)-target
interactions from DoRothEA</p></li>
<li><p><code>tf_target</code>: transcription factor
(TF)-target interactions from other resources</p></li>
<li><p><code>transcriptional</code>: all transcription factor
(TF)-target interactions; this is the combination of the
<em>collectri</em>, <em>dorothea</em> and <em>tf_target</em> datasets.</p></li>
</ul><p><strong>Post-transcriptional (miRNA-target) and other RNA related interactions</strong></p>
<p>In these datasets we intend to collect the literature curated resources,
hence we don't include some of the most well known large databases if those
are based on predictions or high-throughput assays.</p>
<ul><li><p><code>mirna_target</code>: miRNA-mRNA interactions</p></li>
<li><p><code>tf_mirna</code>: TF-miRNA interactions</p></li>
<li><p><code>lncrna_mrna</code>: lncRNA-mRNA interactions</p></li>
</ul><p><strong>Other interaction access functions</strong></p>
<ul><li><p><code>small_molecule</code>: interactions between small molecules and
proteins. Currently this is a small, experimental dataset that
includes drug-target, ligand-receptor, enzyme-metabolite and other
interactions. In the future this will be largely expanded and
divided into multiple datasets.</p></li>
<li><p><code>all_interactions</code>: all the interaction datasets combined.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index">
<ul><li><p><code><a href="interaction_resources.html">interaction_resources</a></code></p></li>
<li><p><code><a href="interaction_graph.html">interaction_graph</a></code></p></li>
<li><p><code><a href="print_interactions.html">print_interactions</a></code></p></li>
<li><p><code><a href="annotated_network.html">annotated_network</a></code></p></li>
</ul><ul><li><p><code>omnipath_interactions</code></p></li>
<li><p><code>post_translational</code></p></li>
<li><p><code><a href="interaction_resources.html">interaction_resources</a></code></p></li>
<li><p><code>all_interactions</code></p></li>
<li><p><code><a href="interaction_graph.html">interaction_graph</a></code></p></li>
<li><p><code><a href="print_interactions.html">print_interactions</a></code></p></li>
</ul></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">op</span> <span class="op">&lt;-</span> <span class="fu">omnipath</span><span class="op">(</span>resources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CA1"</span>, <span class="st">"SIGNOR"</span>, <span class="st">"SignaLink3"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">op</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 63,078 × 15</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    source target source_genesymbol target_genesymbol is_directed is_stimulation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> Q13976 Q13507 PRKG1             TRPC3                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> Q13976 Q9HCX4 PRKG1             TRPC7                       1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> Q13438 Q9HBA0 OS9               TRPV4                       1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> P18031 Q9H1D0 PTPN1             TRPV6                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> P63244 Q9BX84 RACK1             TRPM6                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> P17612 Q9GZU1 PRKACA            MCOLN1                      1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Q16539 P49137 MAPK14            MAPKAPK2                    1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> P49137 Q13151 MAPKAPK2          HNRNPA0                     1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> P49137 O95453 MAPKAPK2          PARN                        1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> O60674 P19235 JAK2              EPOR                        1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 63,068 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 9 more variables: is_inhibition &lt;dbl&gt;, consensus_direction &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   consensus_stimulation &lt;dbl&gt;, consensus_inhibition &lt;dbl&gt;, sources &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   references &lt;chr&gt;, curation_effort &lt;dbl&gt;, n_references &lt;int&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   n_resources &lt;int&gt;</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">interactions</span> <span class="op">=</span> <span class="fu">omnipath_interactions</span><span class="op">(</span></span></span>
<span class="r-in"><span>    resources <span class="op">=</span> <span class="st">"SignaLink3"</span>,</span></span>
<span class="r-in"><span>    organism <span class="op">=</span> <span class="fl">9606</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pathways</span> <span class="op">&lt;-</span> <span class="fu">omnipath</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pathways</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 81,529 × 15</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    source target source_genesymbol target_genesymbol is_directed is_stimulation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> P0DP23 P48995 CALM1             TRPC1                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> P0DP25 P48995 CALM3             TRPC1                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> P0DP24 P48995 CALM2             TRPC1                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> Q03135 P48995 CAV1              TRPC1                       1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> P14416 P48995 DRD2              TRPC1                       1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> Q99750 P48995 MDFI              TRPC1                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Q14571 P48995 ITPR2             TRPC1                       1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> P29966 P48995 MARCKS            TRPC1                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> Q13255 P48995 GRM1              TRPC1                       1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> Q13586 P48995 STIM1             TRPC1                       1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 81,519 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 9 more variables: is_inhibition &lt;dbl&gt;, consensus_direction &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   consensus_stimulation &lt;dbl&gt;, consensus_inhibition &lt;dbl&gt;, sources &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   references &lt;chr&gt;, curation_effort &lt;dbl&gt;, n_references &lt;int&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   n_resources &lt;int&gt;</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">interactions</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>    <span class="fu">pathwayextra</span><span class="op">(</span></span></span>
<span class="r-in"><span>        resources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BioGRID"</span>, <span class="st">"IntAct"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>        organism <span class="op">=</span> <span class="fl">9606</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">kinase_substrate</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>   <span class="fu">kinaseextra</span><span class="op">(</span></span></span>
<span class="r-in"><span>       resources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PhosphoPoint'</span>, <span class="st">'PhosphoSite'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>       organism <span class="op">=</span> <span class="fl">9606</span></span></span>
<span class="r-in"><span>   <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">ligand_receptor</span> <span class="op">&lt;-</span> <span class="fu">ligrecextra</span><span class="op">(</span></span></span>
<span class="r-in"><span>    resources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'HPRD'</span>, <span class="st">'Guide2Pharma'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    organism <span class="op">=</span> <span class="fl">9606</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">interactions</span> <span class="op">&lt;-</span> <span class="fu">post_translational</span><span class="op">(</span>resources <span class="op">=</span> <span class="st">"BioGRID"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">dorothea_grn</span> <span class="op">&lt;-</span> <span class="fu">dorothea</span><span class="op">(</span></span></span>
<span class="r-in"><span>    resources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'DoRothEA'</span>, <span class="st">'ARACNe-GTEx_DoRothEA'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    organism <span class="op">=</span> <span class="fl">9606</span>,</span></span>
<span class="r-in"><span>    dorothea_levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span>, <span class="st">'B'</span>, <span class="st">'C'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dorothea_grn</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 32,617 × 16</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    source target source_genesymbol target_genesymbol is_directed is_stimulation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> P01106 O14746 MYC               TERT                        1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> P84022 P05412 SMAD3             JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> Q13485 P05412 SMAD4             JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> P08047 P04075 SP1               ALDOA                       1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> P04637 P08069 TP53              IGF1R                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> Q05516 P20248 ZBTB16            CCNA2                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Q01196 P08700 RUNX1             IL3                         1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> P42224 P38936 STAT1             CDKN1A                      1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> P40763 P38936 STAT3             CDKN1A                      1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> Q04206 P08183 RELA              ABCB1                       1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 32,607 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 10 more variables: is_inhibition &lt;dbl&gt;, consensus_direction &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   consensus_stimulation &lt;dbl&gt;, consensus_inhibition &lt;dbl&gt;, sources &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   references &lt;chr&gt;, curation_effort &lt;dbl&gt;, dorothea_level &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   n_references &lt;dbl&gt;, n_resources &lt;int&gt;</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">interactions</span> <span class="op">&lt;-</span> <span class="fu">tf_target</span><span class="op">(</span>resources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DoRothEA"</span>, <span class="st">"SIGNOR"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">grn</span> <span class="op">&lt;-</span> <span class="fu">transcriptional</span><span class="op">(</span>resources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PAZAR"</span>, <span class="st">"ORegAnno"</span>, <span class="st">"DoRothEA"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following resources are not available: PAZAR. Check the resource names for spelling mistakes.</span>
<span class="r-in"><span><span class="va">grn</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 31,896 × 16</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    source target source_genesymbol target_genesymbol is_directed is_stimulation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> P01106 O14746 MYC               TERT                        1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> P84022 P05412 SMAD3             JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> Q13485 P05412 SMAD4             JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> Q04206 P25445 RELA              FAS                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> P08047 P04075 SP1               ALDOA                       1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> P04637 P08069 TP53              IGF1R                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Q05516 P20248 ZBTB16            CCNA2                       1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> Q01196 P08700 RUNX1             IL3                         1              0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> P42224 P38936 STAT1             CDKN1A                      1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> P40763 P38936 STAT3             CDKN1A                      1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 31,886 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 10 more variables: is_inhibition &lt;dbl&gt;, consensus_direction &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   consensus_stimulation &lt;dbl&gt;, consensus_inhibition &lt;dbl&gt;, sources &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   references &lt;chr&gt;, curation_effort &lt;dbl&gt;, dorothea_level &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   n_references &lt;dbl&gt;, n_resources &lt;int&gt;</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">collectri_grn</span> <span class="op">&lt;-</span> <span class="fu">collectri</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">collectri_grn</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 64,495 × 15</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    source  target source_genesymbol target_genesymbol is_directed is_stimulation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> P01106  O14746 MYC               TERT                        1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> P17947  P02818 SPI1              BGLAP                       1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> COMPLE… P05412 FOSL1_JUNB        JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> COMPLE… P05412 FOS_JUN           JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> COMPLE… P05412 FOS_JUNB          JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> COMPLE… P05412 FOSL2_JUND        JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> COMPLE… P05412 FOSL2_JUN         JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> COMPLE… P05412 JUN               JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> COMPLE… P05412 FOSB_JUNB         JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> COMPLE… P05412 JUNB              JUN                         1              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 64,485 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 9 more variables: is_inhibition &lt;dbl&gt;, consensus_direction &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   consensus_stimulation &lt;dbl&gt;, consensus_inhibition &lt;dbl&gt;, sources &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   references &lt;chr&gt;, curation_effort &lt;dbl&gt;, n_references &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   n_resources &lt;int&gt;</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">interactions</span> <span class="op">&lt;-</span> <span class="fu">mirna_target</span><span class="op">(</span> resources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"miRTarBase"</span>, <span class="st">"miRecords"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">interactions</span> <span class="op">&lt;-</span> <span class="fu">tf_mirna</span><span class="op">(</span>resources <span class="op">=</span> <span class="st">"TransmiR"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">interactions</span> <span class="op">&lt;-</span> <span class="fu">lncrna_mrna</span><span class="op">(</span>resources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ncRDeathDB"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># What are the targets of aspirin?</span></span></span>
<span class="r-in"><span><span class="va">interactions</span> <span class="op">&lt;-</span> <span class="fu">small_molecule</span><span class="op">(</span>sources <span class="op">=</span> <span class="st">"ASPIRIN"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># The prostaglandin synthases:</span></span></span>
<span class="r-in"><span><span class="va">interactions</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 0 × 14</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 14 variables: source &lt;chr&gt;, target &lt;chr&gt;, source_genesymbol &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   target_genesymbol &lt;chr&gt;, is_directed &lt;chr&gt;, is_stimulation &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   is_inhibition &lt;chr&gt;, consensus_direction &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   consensus_stimulation &lt;chr&gt;, consensus_inhibition &lt;chr&gt;, sources &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   references &lt;chr&gt;, curation_effort &lt;chr&gt;, n_references &lt;lgl&gt;</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">interactions</span> <span class="op">&lt;-</span> <span class="fu">all_interactions</span><span class="op">(</span></span></span>
<span class="r-in"><span>    resources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HPRD"</span>, <span class="st">"BioGRID"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    organism <span class="op">=</span> <span class="fl">9606</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Valdeolivas, Denes Turei, Attila Gabor, Diego Mananes, Aurelien Dugourd.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

