<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Translate gene, protein and small molecule identifiers — translate_ids • OmnipathR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Translate gene, protein and small molecule identifiers — translate_ids"><meta name="description" content="Translates a vector of identifiers, resulting a new vector, or a column
of identifiers in a data frame by creating another column with the target
identifiers."><meta property="og:description" content="Translates a vector of identifiers, resulting a new vector, or a column
of identifiers in a data frame by creating another column with the target
identifiers."><meta property="og:image" content="https://saezlab.github.io/OmnipathR/logo.svg"><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-10"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-10');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">OmnipathR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.15.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bioc_workshop.html">OmniPath Bioconductor workshop</a></li>
    <li><a class="dropdown-item" href="../articles/cosmos.html">Building prior knowledge network (PKN) for COSMOS</a></li>
    <li><a class="dropdown-item" href="../articles/db_manager.html">The database manager in OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/drug_targets.html">Building protein networks around drug-targets using OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/extra_attrs.html">Resource specific interaction attributes</a></li>
    <li><a class="dropdown-item" href="../articles/gene_ontology.html">Gene Ontology in OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/nichenet.html">Using NicheNet with OmnipathR</a></li>
    <li><a class="dropdown-item" href="../articles/omnipath_intro.html">OmnipathR: an R client for the OmniPath web service</a></li>
    <li><a class="dropdown-item" href="../articles/paths.html">Pathway construction</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/saezlab/OmnipathR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/omnipathdb" aria-label="Go to the Twitter of OmniPath"><span class="fa fab fa-twitter fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://omnipathdb.org/" aria-label="Go to the OmniPath homepage"><span class="fa fas fa-university"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Translate gene, protein and small molecule identifiers</h1>
      <small class="dont-index">Source: <a href="https://github.com/saezlab/OmnipathR/blob/HEAD/R/id_mapping.R" class="external-link"><code>R/id_mapping.R</code></a></small>
      <div class="d-none name"><code>translate_ids.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Translates a vector of identifiers, resulting a new vector, or a column
of identifiers in a data frame by creating another column with the target
identifiers.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">translate_ids</span><span class="op">(</span></span>
<span>  <span class="va">d</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  uploadlists <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ensembl <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  hmdb <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ramp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chalmers <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  entity_type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  keep_untranslated <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_df <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  organism <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>  reviewed <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  complexes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  complexes_one_to_many <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  track <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  quantify_ambiguity <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  qualify_ambiguity <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ambiguity_groups <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ambiguity_global <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ambiguity_summary <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  expand <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-d">d<a class="anchor" aria-label="anchor" href="#arg-d"></a></dt>
<dd><p>Character vector or data frame.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>At least two arguments, with or without names. The first
of these arguments describes the source identifier, the rest
of them describe the target identifier(s). The values of all these
arguments must be valid identifier types as shown in Details. The
names of the arguments are column names. In case of the first
(source) ID the column must exist. For the rest of the IDs new
columns will be created with the desired names. For ID types provided
as arguments without names, the name of the ID type will be used for
column name.</p></dd>


<dt id="arg-uploadlists">uploadlists<a class="anchor" aria-label="anchor" href="#arg-uploadlists"></a></dt>
<dd><p>Force using the <code>uploadlists</code> service from UniProt.
By default the plain query interface is used (implemented in
<code><a href="uniprot_full_id_mapping_table.html">uniprot_full_id_mapping_table</a></code> in this package).
If any of the provided ID types is only available in the uploadlists
service, it will be automatically selected. The plain query interface
is preferred because in the long term, with caching, it requires
less download and data storage.</p></dd>


<dt id="arg-ensembl">ensembl<a class="anchor" aria-label="anchor" href="#arg-ensembl"></a></dt>
<dd><p>Logical: use data from Ensembl BioMart instead of UniProt.</p></dd>


<dt id="arg-hmdb">hmdb<a class="anchor" aria-label="anchor" href="#arg-hmdb"></a></dt>
<dd><p>Logical: use HMDB ID translation data.</p></dd>


<dt id="arg-ramp">ramp<a class="anchor" aria-label="anchor" href="#arg-ramp"></a></dt>
<dd><p>Logical: use RaMP ID translation data.</p></dd>


<dt id="arg-chalmers">chalmers<a class="anchor" aria-label="anchor" href="#arg-chalmers"></a></dt>
<dd><p>Logical: use ID translation data from Chalmers Sysbio GEM.</p></dd>


<dt id="arg-entity-type">entity_type<a class="anchor" aria-label="anchor" href="#arg-entity-type"></a></dt>
<dd><p>Character: "gene" and "smol" are short symbols for
proteins, genes and small molecules respectively. Several other synonyms
are also accepted.</p></dd>


<dt id="arg-keep-untranslated">keep_untranslated<a class="anchor" aria-label="anchor" href="#arg-keep-untranslated"></a></dt>
<dd><p>In case the output is a data frame, keep the
records where the source identifier could not be translated. At
these records the target identifier will be NA.</p></dd>


<dt id="arg-return-df">return_df<a class="anchor" aria-label="anchor" href="#arg-return-df"></a></dt>
<dd><p>Return a data frame even if the input is a vector.</p></dd>


<dt id="arg-organism">organism<a class="anchor" aria-label="anchor" href="#arg-organism"></a></dt>
<dd><p>Character or integer, name or NCBI Taxonomy ID of the
organism (by default 9606 for human). Matters only if
<code>uploadlists</code> is <code>FALSE</code>.</p></dd>


<dt id="arg-reviewed">reviewed<a class="anchor" aria-label="anchor" href="#arg-reviewed"></a></dt>
<dd><p>Translate only reviewed (<code>TRUE</code>), only unreviewed
(<code>FALSE</code>) or both (<code>NULL</code>) UniProt records. Matters only
if <code>uploadlists</code> is <code>FALSE</code>.</p></dd>


<dt id="arg-complexes">complexes<a class="anchor" aria-label="anchor" href="#arg-complexes"></a></dt>
<dd><p>Logical: translate complexes by their members. Only
complexes where all members can be translated will be included in the
result. If <code>NULL</code>, the option <code>omnipathr.complex_translation</code> will
be used.</p></dd>


<dt id="arg-complexes-one-to-many">complexes_one_to_many<a class="anchor" aria-label="anchor" href="#arg-complexes-one-to-many"></a></dt>
<dd><p>Logical: allow combinatorial expansion or
use only the first target identifier for each member of each complex.
If <code>NULL</code>, the option <code>omnipathr.complex_translation_one_to_many</code>
will be used.</p></dd>


<dt id="arg-track">track<a class="anchor" aria-label="anchor" href="#arg-track"></a></dt>
<dd><p>Logical: Track the records (rows) in the input data frame by
adding a column <code>record_id</code> with the original row numbers.</p></dd>


<dt id="arg-quantify-ambiguity">quantify_ambiguity<a class="anchor" aria-label="anchor" href="#arg-quantify-ambiguity"></a></dt>
<dd><p>Logical or character: inspect the mappings for each
ID for ambiguity. If TRUE, for each translated column, two new columns
will be created with numeric values, representing the ambiguity of the
mapping on the "from" and "to" side of the translation, respectively.
If a character value provided, it will be used as a column name suffix
for the new columns.</p></dd>


<dt id="arg-qualify-ambiguity">qualify_ambiguity<a class="anchor" aria-label="anchor" href="#arg-qualify-ambiguity"></a></dt>
<dd><p>Logical or character: inspect the mappings for each
ID for ambiguity. If TRUE, for each translated column, a new column
will be inculded with values <code>one-to-one</code>, <code>one-to-many</code>, <code>many-to-one</code>
or <code>many-to-many</code>. If a character value provided, it will be used as a
column name suffix for the new column.</p></dd>


<dt id="arg-ambiguity-groups">ambiguity_groups<a class="anchor" aria-label="anchor" href="#arg-ambiguity-groups"></a></dt>
<dd><p>Character vector: additional column names to group by
during inspecting ambiguity. By default, the identifier columns (from
and to) will be used to determine the ambiguity of mappings.</p></dd>


<dt id="arg-ambiguity-global">ambiguity_global<a class="anchor" aria-label="anchor" href="#arg-ambiguity-global"></a></dt>
<dd><p>Logical or character: if <code>ambiguity_groups</code> are
provided, analyse ambiguity also globally, across the whole data frame.
Character value provides a custom suffix for the columns quantifying
and qualifying global ambiguity.</p></dd>


<dt id="arg-ambiguity-summary">ambiguity_summary<a class="anchor" aria-label="anchor" href="#arg-ambiguity-summary"></a></dt>
<dd><p>Logical: generate a summary about the ambiguity of the
translation and make it available as an attribute.
columns will be lists of character vectors.</p></dd>


<dt id="arg-expand">expand<a class="anchor" aria-label="anchor" href="#arg-expand"></a></dt>
<dd><p>Logical: if <code>TRUE</code>, ambiguous (to-many) mappings will be
expanded to multiple rows, resulting character type columns; if
<code>FALSE</code>, the original rows will be kept intact, and the target</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>

<ul><li><p>Data frame: if the input is a data frame or the input is a
vector and <code>return_df</code> is <code>TRUE</code>.</p></li>
<li><p>Vector: if the input is a vector, there is only one target
ID type and <code>return_df</code> is <code>FALSE</code>.</p></li>
<li><p>List of vectors: if the input is a vector, there are more than
one target ID types and <code>return_df</code> is <code>FALSE</code>. The names
of the list will be ID types (as they were column names, see
the description of the <code>...</code> argument), and the list will also
include the source IDs.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function, depending on the <code>uploadlists</code> parameter, uses either
the uploadlists service of UniProt or plain UniProt queries to obtain
identifier translation tables. The possible values for <code>from</code> and <code>to</code>
are the identifier type abbreviations used in the UniProt API, please
refer to the table here: <a href="https://www.uniprot.org/help/api_idmapping" class="external-link">https://www.uniprot.org/help/api_idmapping</a>.
In addition, simple synonyms are available which realize a uniform API
for the uploadlists and UniProt query based backends. These are the
followings:</p><table class="table table"><tr><td><strong>OmnipathR</strong></td><td><strong>Uploadlists</strong></td><td><strong>UniProt query</strong></td><td><strong>Ensembl BioMart</strong></td></tr><tr><td>uniprot</td><td>ACC</td><td>id</td><td>uniprotswissprot</td></tr><tr><td>uniprot_entry</td><td>ID</td><td>entry name</td><td></td></tr><tr><td>trembl</td><td><em>reviewed = FALSE</em></td><td><em>reviewed = FALSE</em></td><td>uniprotsptrembl</td></tr><tr><td>genesymbol</td><td>GENENAME</td><td>genes(PREFERRED)</td><td>external_gene_name</td></tr><tr><td>genesymbol_syn</td><td></td><td>genes(ALTERNATIVE)</td><td>external_synonym</td></tr><tr><td>hgnc</td><td>HGNC_ID</td><td>database(HGNC)</td><td>hgnc_symbol</td></tr><tr><td>entrez</td><td>P_ENTREZGENEID</td><td>database(GeneID)</td><td></td></tr><tr><td>ensembl</td><td>ENSEMBL_ID</td><td></td><td>ensembl_gene_id</td></tr><tr><td>ensg</td><td>ENSEMBL_ID</td><td></td><td>ensembl_gene_id</td></tr><tr><td>enst</td><td>ENSEMBL_TRS_ID</td><td>database(Ensembl)</td><td>ensembl_transcript_id</td></tr><tr><td>ensp</td><td>ENSEMBL_PRO_ID</td><td></td><td>ensembl_peptide_id</td></tr><tr><td>ensgg</td><td>ENSEMBLGENOME_ID</td><td></td><td></td></tr><tr><td>ensgt</td><td>ENSEMBLGENOME_TRS_ID</td><td></td><td></td></tr><tr><td>ensgp</td><td>ENSEMBLGENOME_PRO_ID</td><td></td><td></td></tr><tr><td>protein_name</td><td></td><td>protein names</td><td></td></tr><tr><td>pir</td><td>PIR</td><td>database(PIR)</td><td></td></tr><tr><td>ccds</td><td></td><td>database(CCDS)</td><td></td></tr><tr><td>refseqp</td><td>P_REFSEQ_AC</td><td>database(refseq)</td><td></td></tr><tr><td>ipro</td><td></td><td></td><td>interpro</td></tr><tr><td>ipro_desc</td><td></td><td></td><td>interpro_description</td></tr><tr><td>ipro_sdesc</td><td></td><td></td><td>interpro_short_description</td></tr><tr><td>wikigene</td><td></td><td></td><td>wikigene_name</td></tr><tr><td>rnacentral</td><td></td><td></td><td>rnacentral</td></tr><tr><td>gene_desc</td><td></td><td></td><td>description</td></tr><tr><td>wormbase</td><td></td><td>database(WormBase)</td><td></td></tr><tr><td>flybase</td><td></td><td>database(FlyBase)</td><td></td></tr><tr><td>xenbase</td><td></td><td>database(Xenbase)</td><td></td></tr><tr><td>zfin</td><td></td><td>database(ZFIN)</td><td></td></tr><tr><td>pbd</td><td>PBD_ID</td><td>database(PDB)</td><td>pbd</td></tr></table><p>For a complete list of ID types and their synonyms, including metabolite and
chemical ID types which are not shown here, see <code><a href="id_types.html">id_types</a></code>.</p>
<p>The mapping between identifiers can be ambiguous. In this case one row
in the original data frame yields multiple rows or elements in the
returned data frame or vector(s).</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index">
<ul><li><p><code><a href="translate_ids_multi.html">translate_ids_multi</a></code></p></li>
<li><p><code><a href="uniprot_id_mapping_table.html">uniprot_id_mapping_table</a></code></p></li>
<li><p><code><a href="uniprot_full_id_mapping_table.html">uniprot_full_id_mapping_table</a></code></p></li>
<li><p><code><a href="ensembl_id_mapping_table.html">ensembl_id_mapping_table</a></code></p></li>
<li><p><code><a href="hmdb_id_mapping_table.html">hmdb_id_mapping_table</a></code></p></li>
<li><p><code><a href="id_types.html">id_types</a></code></p></li>
<li><p><code><a href="ensembl_id_type.html">ensembl_id_type</a></code></p></li>
<li><p><code><a href="uniprot_id_type.html">uniprot_id_type</a></code></p></li>
<li><p><code><a href="uploadlists_id_type.html">uploadlists_id_type</a></code></p></li>
<li><p><code><a href="hmdb_id_type.html">hmdb_id_type</a></code></p></li>
<li><p><code><a href="chalmers_gem_id_type.html">chalmers_gem_id_type</a></code></p></li>
</ul></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    uniprot_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>        <span class="st">'P00533'</span>, <span class="st">'Q9ULV1'</span>, <span class="st">'P43897'</span>, <span class="st">'Q9Y2P5'</span>,</span></span>
<span class="r-in"><span>        <span class="st">'P01258'</span>, <span class="st">'P06881'</span>, <span class="st">'P42771'</span>, <span class="st">'Q8N726'</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu">translate_ids</span><span class="op">(</span><span class="va">d</span>, uniprot_id <span class="op">=</span> <span class="va">uniprot</span>, <span class="va">genesymbol</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">d</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   uniprot_id genesymbol</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     P00533       EGFR</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     Q9ULV1       FZD4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     P43897       TSFM</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     Q9Y2P5    SLC27A5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5     P01258      CALCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6     P06881      CALCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7     P42771     CDKN2A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8     Q8N726     CDKN2A</span>
<span class="r-in"><span><span class="co">#   uniprot_id genesymbol</span></span></span>
<span class="r-in"><span><span class="co"># 1     P00533       EGFR</span></span></span>
<span class="r-in"><span><span class="co"># 2     Q9ULV1       FZD4</span></span></span>
<span class="r-in"><span><span class="co"># 3     P43897       TSFM</span></span></span>
<span class="r-in"><span><span class="co"># 4     Q9Y2P5    SLC27A5</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Valdeolivas, Denes Turei, Attila Gabor, Diego Mananes, Aurelien Dugourd.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

